<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.30">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Bill Cresko">

<title>Week 8b - Statistics for Bioengineering – BioEngineering Statistics Winter 2026</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-de070a7b0ab54f8780927367ac907214.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-f7fe251563e67c05764dc4775ff48cfc.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">BioEngineering Statistics Winter 2026</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../Syllabus.html"> 
<span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../Policies.html"> 
<span class="menu-text">Policies</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../Schedule.html"> 
<span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../Lectures.html"> 
<span class="menu-text">Lectures</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../Homework.html"> 
<span class="menu-text">Homeworks</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../Software.html"> 
<span class="menu-text">Software</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../Resources.html"> 
<span class="menu-text">Resources</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#this-week" id="toc-this-week" class="nav-link active" data-scroll-target="#this-week">This week</a></li>
  <li><a href="#variance-components-with-2-random-factors-using-lme4" id="toc-variance-components-with-2-random-factors-using-lme4" class="nav-link" data-scroll-target="#variance-components-with-2-random-factors-using-lme4">Variance components with 2 random factors using LME4</a></li>
  <li><a href="#generalized-linear-models" id="toc-generalized-linear-models" class="nav-link" data-scroll-target="#generalized-linear-models">Generalized Linear Models</a>
  <ul class="collapse">
  <li><a href="#using-glm-for-more-complicated-mixed-models" id="toc-using-glm-for-more-complicated-mixed-models" class="nav-link" data-scroll-target="#using-glm-for-more-complicated-mixed-models">Using <code>glm</code> for more complicated mixed models</a></li>
  <li><a href="#a-generalized-linear-model" id="toc-a-generalized-linear-model" class="nav-link" data-scroll-target="#a-generalized-linear-model">A Generalized Linear Model</a></li>
  <li><a href="#glm-to-test-wolbachia-infection" id="toc-glm-to-test-wolbachia-infection" class="nav-link" data-scroll-target="#glm-to-test-wolbachia-infection"><code>glm</code> to test Wolbachia infection</a></li>
  <li><a href="#adding-more-variables" id="toc-adding-more-variables" class="nav-link" data-scroll-target="#adding-more-variables">Adding more variables</a></li>
  <li><a href="#adding-interaction-effects" id="toc-adding-interaction-effects" class="nav-link" data-scroll-target="#adding-interaction-effects">Adding interaction effects</a></li>
  <li><a href="#how-do-we-know-which-model-to-use" id="toc-how-do-we-know-which-model-to-use" class="nav-link" data-scroll-target="#how-do-we-know-which-model-to-use">How do we know which model to use?</a></li>
  <li><a href="#what-is-deviance" id="toc-what-is-deviance" class="nav-link" data-scroll-target="#what-is-deviance">What is Deviance?</a></li>
  <li><a href="#mathematical-definition" id="toc-mathematical-definition" class="nav-link" data-scroll-target="#mathematical-definition">Mathematical Definition</a></li>
  <li><a href="#types-of-deviance-in-glm" id="toc-types-of-deviance-in-glm" class="nav-link" data-scroll-target="#types-of-deviance-in-glm">Types of Deviance in <code>glm()</code></a></li>
  <li><a href="#interpreting-deviance" id="toc-interpreting-deviance" class="nav-link" data-scroll-target="#interpreting-deviance">Interpreting Deviance</a></li>
  <li><a href="#example-output" id="toc-example-output" class="nav-link" data-scroll-target="#example-output">Example Output</a></li>
  </ul></li>
  <li><a href="#generalized-linear-models-1" id="toc-generalized-linear-models-1" class="nav-link" data-scroll-target="#generalized-linear-models-1">Generalized Linear Models</a>
  <ul class="collapse">
  <li><a href="#what-is-a-generalized-linear-model" id="toc-what-is-a-generalized-linear-model" class="nav-link" data-scroll-target="#what-is-a-generalized-linear-model">What is a generalized linear model?</a></li>
  <li><a href="#generalized-linear-models-have-three-components" id="toc-generalized-linear-models-have-three-components" class="nav-link" data-scroll-target="#generalized-linear-models-have-three-components">Generalized linear models have three components</a></li>
  <li><a href="#assumptions" id="toc-assumptions" class="nav-link" data-scroll-target="#assumptions">Assumptions</a></li>
  </ul></li>
  <li><a href="#logistic-regression" id="toc-logistic-regression" class="nav-link" data-scroll-target="#logistic-regression">Logistic Regression</a>
  <ul class="collapse">
  <li><a href="#what-is-logistic-regression" id="toc-what-is-logistic-regression" class="nav-link" data-scroll-target="#what-is-logistic-regression">What is Logistic Regression?</a></li>
  <li><a href="#why-not-linear-regression" id="toc-why-not-linear-regression" class="nav-link" data-scroll-target="#why-not-linear-regression">Why Not Linear Regression?</a></li>
  <li><a href="#the-logistic-function" id="toc-the-logistic-function" class="nav-link" data-scroll-target="#the-logistic-function">The Logistic Function</a></li>
  <li><a href="#odds-and-log-odds" id="toc-odds-and-log-odds" class="nav-link" data-scroll-target="#odds-and-log-odds">Odds and Log-Odds</a></li>
  <li><a href="#equivalent-model-equation" id="toc-equivalent-model-equation" class="nav-link" data-scroll-target="#equivalent-model-equation">Equivalent Model Equation</a></li>
  <li><a href="#fitting-the-model" id="toc-fitting-the-model" class="nav-link" data-scroll-target="#fitting-the-model">Fitting the Model</a></li>
  <li><a href="#interpreting-coefficients" id="toc-interpreting-coefficients" class="nav-link" data-scroll-target="#interpreting-coefficients">Interpreting Coefficients</a></li>
  <li><a href="#making-predictions" id="toc-making-predictions" class="nav-link" data-scroll-target="#making-predictions">Making Predictions</a></li>
  <li><a href="#model-evaluation-metrics" id="toc-model-evaluation-metrics" class="nav-link" data-scroll-target="#model-evaluation-metrics">Model Evaluation Metrics</a></li>
  <li><a href="#confusion-matrix" id="toc-confusion-matrix" class="nav-link" data-scroll-target="#confusion-matrix">Confusion Matrix</a></li>
  <li><a href="#accuracy" id="toc-accuracy" class="nav-link" data-scroll-target="#accuracy">Accuracy</a></li>
  <li><a href="#precision" id="toc-precision" class="nav-link" data-scroll-target="#precision">Precision</a></li>
  <li><a href="#recall-sensitivity" id="toc-recall-sensitivity" class="nav-link" data-scroll-target="#recall-sensitivity">Recall (Sensitivity)</a></li>
  <li><a href="#precision-vs.-recall" id="toc-precision-vs.-recall" class="nav-link" data-scroll-target="#precision-vs.-recall">Precision vs.&nbsp;Recall</a></li>
  <li><a href="#auc-roc" id="toc-auc-roc" class="nav-link" data-scroll-target="#auc-roc">AUC-ROC</a></li>
  <li><a href="#r-interlude-logistic-regression" id="toc-r-interlude-logistic-regression" class="nav-link" data-scroll-target="#r-interlude-logistic-regression">R INTERLUDE | Logistic regression</a></li>
  <li><a href="#dataset-variable-descriptions" id="toc-dataset-variable-descriptions" class="nav-link" data-scroll-target="#dataset-variable-descriptions">Dataset Variable Descriptions</a></li>
  <li><a href="#fitting-the-model-1" id="toc-fitting-the-model-1" class="nav-link" data-scroll-target="#fitting-the-model-1">Fitting the model</a></li>
  <li><a href="#fitting-the-model-2" id="toc-fitting-the-model-2" class="nav-link" data-scroll-target="#fitting-the-model-2">Fitting the model</a></li>
  <li><a href="#plotting-the-model" id="toc-plotting-the-model" class="nav-link" data-scroll-target="#plotting-the-model">Plotting the model</a></li>
  </ul></li>
  <li><a href="#poisson-regression" id="toc-poisson-regression" class="nav-link" data-scroll-target="#poisson-regression">Poisson Regression</a>
  <ul class="collapse">
  <li><a href="#what-is-poisson-regression" id="toc-what-is-poisson-regression" class="nav-link" data-scroll-target="#what-is-poisson-regression">What is Poisson Regression?</a></li>
  <li><a href="#the-poisson-distribution" id="toc-the-poisson-distribution" class="nav-link" data-scroll-target="#the-poisson-distribution">The Poisson Distribution</a></li>
  <li><a href="#model-structure" id="toc-model-structure" class="nav-link" data-scroll-target="#model-structure">Model Structure</a></li>
  <li><a href="#why-not-use-linear-regression" id="toc-why-not-use-linear-regression" class="nav-link" data-scroll-target="#why-not-use-linear-regression">Why Not Use Linear Regression?</a></li>
  <li><a href="#example" id="toc-example" class="nav-link" data-scroll-target="#example">Example</a></li>
  <li><a href="#interpreting-coefficients-1" id="toc-interpreting-coefficients-1" class="nav-link" data-scroll-target="#interpreting-coefficients-1">Interpreting Coefficients</a></li>
  <li><a href="#model-assumptions" id="toc-model-assumptions" class="nav-link" data-scroll-target="#model-assumptions">Model Assumptions</a></li>
  <li><a href="#model-evaluation" id="toc-model-evaluation" class="nav-link" data-scroll-target="#model-evaluation">Model Evaluation</a></li>
  <li><a href="#overdispersion" id="toc-overdispersion" class="nav-link" data-scroll-target="#overdispersion">Overdispersion</a></li>
  <li><a href="#goodness-of-fit" id="toc-goodness-of-fit" class="nav-link" data-scroll-target="#goodness-of-fit">Goodness-of-Fit</a></li>
  <li><a href="#r-interlude-log-linear-poisson-regression" id="toc-r-interlude-log-linear-poisson-regression" class="nav-link" data-scroll-target="#r-interlude-log-linear-poisson-regression">R INTERLUDE | Log-linear (Poisson) regression</a></li>
  <li><a href="#dataset-variable-descriptions-poisson-model" id="toc-dataset-variable-descriptions-poisson-model" class="nav-link" data-scroll-target="#dataset-variable-descriptions-poisson-model">Dataset Variable Descriptions (Poisson Model)</a></li>
  <li><a href="#fit-the-model" id="toc-fit-the-model" class="nav-link" data-scroll-target="#fit-the-model">Fit the model</a></li>
  <li><a href="#fit-the-model-1" id="toc-fit-the-model-1" class="nav-link" data-scroll-target="#fit-the-model-1">Fit the model</a></li>
  <li><a href="#make-a-plot" id="toc-make-a-plot" class="nav-link" data-scroll-target="#make-a-plot">Make a plot</a></li>
  </ul></li>
  <li><a href="#section" id="toc-section" class="nav-link" data-scroll-target="#section"></a></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="Week8b.html"><i class="bi bi-file-slides"></i>RevealJS</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Week 8b - Statistics for Bioengineering</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Bill Cresko </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'ggplot2' was built under R version 4.4.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'tidyr' was built under R version 4.4.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'readr' was built under R version 4.4.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.6
✔ forcats   1.0.1     ✔ stringr   1.6.0
✔ ggplot2   4.0.1     ✔ tibble    3.3.0
✔ lubridate 1.9.4     ✔ tidyr     1.3.2
✔ purrr     1.2.0     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'gt' was built under R version 4.4.3</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="this-week" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="this-week">This week</h2>
<ul>
<li><p>Key principles of experimental design</p></li>
<li><p>Linear Mixed Models (GLMMs)</p></li>
<li><p>Generalized Linear Models - Logistic and Poisson Regression</p></li>
<li><p>Chi-square tests of association and goodness of fit</p></li>
</ul>
</section>
<section id="variance-components-with-2-random-factors-using-lme4" class="level1 smaller">
<h1 class="smaller">Variance components with 2 random factors using LME4</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>rnadata <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">'Stickle_RNAseq.tsv'</span>, <span class="at">header=</span>T, <span class="at">sep=</span><span class="st">''</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>gene <span class="ot">&lt;-</span> rnadata<span class="sc">$</span>Gene80[<span class="dv">6</span><span class="sc">:</span><span class="dv">75</span>] </span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>microbiota <span class="ot">&lt;-</span> rnadata<span class="sc">$</span>Microbiota[<span class="dv">6</span><span class="sc">:</span><span class="dv">75</span>]</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>genotype <span class="ot">&lt;-</span> rnadata<span class="sc">$</span>Genotype[<span class="dv">6</span><span class="sc">:</span><span class="dv">75</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Estimate the variance components using Restricted Maximum Likelihood (REML)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(lme4)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'lme4' was built under R version 4.4.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: Matrix</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'Matrix'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:tidyr':

    expand, pack, unpack</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">lmer</span>(gene <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> microbiota) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> genotype) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> microbiota<span class="sc">:</span>genotype))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Linear mixed model fit by REML ['lmerMod']
Formula: 
gene ~ 1 + (1 | microbiota) + (1 | genotype) + (1 | microbiota:genotype)
REML criterion at convergence: 865.1191
Random effects:
 Groups              Name        Std.Dev.
 microbiota:genotype (Intercept)  65.03  
 genotype            (Intercept)  34.41  
 microbiota          (Intercept) 118.60  
 Residual                        117.06  
Number of obs: 70, groups:  microbiota:genotype, 4; genotype, 2; microbiota, 2
Fixed Effects:
(Intercept)  
      180.8  </code></pre>
</div>
</div>
<hr>
</section>
<section id="generalized-linear-models" class="level1">
<h1>Generalized Linear Models</h1>
<section id="using-glm-for-more-complicated-mixed-models" class="level2">
<h2 class="anchored" data-anchor-id="using-glm-for-more-complicated-mixed-models">Using <code>glm</code> for more complicated mixed models</h2>
<ul>
<li>For a single categorical predictor, we can include effects of each factor level:</li>
</ul>
<p><span class="math display">\[ y_i = \beta_0 + \beta_1(x~level1) + \beta_2(x~level2) + \beta_3x_3~... ~+ e_1 \]</span></p>
<ul>
<li>Each factor level becomes it’s own “effector” that will change the value of y</li>
<li>Instead of x being a continuous numerical value, for categorical data it will be either a 0 or 1</li>
<li><span class="math inline">\(x_3\)</span> is also a continuous variable so this is a mixed model</li>
</ul>
</section>
<section id="a-generalized-linear-model" class="level2">
<h2 class="anchored" data-anchor-id="a-generalized-linear-model">A Generalized Linear Model</h2>
<ul>
<li>Check out the help page in RStudio for <code>glm()</code></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">glm</span>(<span class="at">formula =</span> y <span class="sc">~</span> x1 <span class="sc">+</span> x2 <span class="sc">+</span> x3, <span class="at">family =</span> <span class="fu">gaussian</span>(<span class="at">link =</span> <span class="st">"identity"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="glm-to-test-wolbachia-infection" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="glm-to-test-wolbachia-infection"><code>glm</code> to test Wolbachia infection</h2>
<ul>
<li>Set up a <code>glm()</code> to test for the effect of Wolbachia infection on Recombinant Fraction</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>fly <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Mostoufi2022_Recombination_Edit.csv"</span>, <span class="at">header=</span>T, <span class="at">sep=</span><span class="st">','</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">glm</span>(fly<span class="sc">$</span>RecombinantFraction <span class="sc">~</span> fly<span class="sc">$</span>Wolbachia, <span class="at">family =</span> <span class="fu">gaussian</span>(<span class="at">link=</span><span class="st">"identity"</span>))</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = fly$RecombinantFraction ~ fly$Wolbachia, family = gaussian(link = "identity"))

Coefficients:
                 Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)      0.348054   0.005845  59.545   &lt;2e-16 ***
fly$WolbachiaYes 0.021664   0.008474   2.557   0.0112 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for gaussian family taken to be 0.00406577)

    Null deviance: 0.94137  on 226  degrees of freedom
Residual deviance: 0.91480  on 225  degrees of freedom
AIC: -601.48

Number of Fisher Scoring iterations: 2</code></pre>
</div>
</div>
</section>
<section id="adding-more-variables" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="adding-more-variables">Adding more variables</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>model2 <span class="ot">&lt;-</span> <span class="fu">glm</span>(fly<span class="sc">$</span>RecombinantFraction <span class="sc">~</span> fly<span class="sc">$</span>Wolbachia <span class="sc">+</span> fly<span class="sc">$</span>Food, <span class="at">family =</span> <span class="fu">gaussian</span>(<span class="at">link =</span> <span class="st">"identity"</span>))</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = fly$RecombinantFraction ~ fly$Wolbachia + fly$Food, 
    family = gaussian(link = "identity"))

Coefficients:
                 Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)      0.344691   0.008121  42.445   &lt;2e-16 ***
fly$WolbachiaYes 0.021731   0.008504   2.555   0.0113 *  
fly$FoodSucrose  0.002933   0.010481   0.280   0.7799    
fly$FoodYeast    0.007366   0.010128   0.727   0.4678    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for gaussian family taken to be 0.00409246)

    Null deviance: 0.94137  on 226  degrees of freedom
Residual deviance: 0.91262  on 223  degrees of freedom
AIC: -598.02

Number of Fisher Scoring iterations: 2</code></pre>
</div>
</div>
</section>
<section id="adding-interaction-effects" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="adding-interaction-effects">Adding interaction effects</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>model3 <span class="ot">&lt;-</span> <span class="fu">glm</span>(fly<span class="sc">$</span>RecombinantFraction <span class="sc">~</span> fly<span class="sc">$</span>Wolbachia <span class="sc">+</span> fly<span class="sc">$</span>Food <span class="sc">+</span> fly<span class="sc">$</span>Wolbachia<span class="sc">*</span>fly<span class="sc">$</span>Food, <span class="at">family =</span> <span class="fu">gaussian</span>(<span class="at">link =</span> <span class="st">"identity"</span>))</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model3)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = fly$RecombinantFraction ~ fly$Wolbachia + fly$Food + 
    fly$Wolbachia * fly$Food, family = gaussian(link = "identity"))

Coefficients:
                                  Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)                       0.344520   0.009760  35.298   &lt;2e-16 ***
fly$WolbachiaYes                  0.022080   0.013970   1.581    0.115    
fly$FoodSucrose                   0.010286   0.014458   0.711    0.478    
fly$FoodYeast                     0.001256   0.014059   0.089    0.929    
fly$WolbachiaYes:fly$FoodSucrose -0.015873   0.021002  -0.756    0.451    
fly$WolbachiaYes:fly$FoodYeast    0.012910   0.020282   0.637    0.525    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for gaussian family taken to be 0.004096248)

    Null deviance: 0.94137  on 226  degrees of freedom
Residual deviance: 0.90527  on 221  degrees of freedom
AIC: -595.86

Number of Fisher Scoring iterations: 2</code></pre>
</div>
</div>
</section>
<section id="how-do-we-know-which-model-to-use" class="level2">
<h2 class="anchored" data-anchor-id="how-do-we-know-which-model-to-use">How do we know which model to use?</h2>
<ul>
<li>Want to be careful not to <em>overfit</em> your linear model (give too many variables)</li>
<li>We can use Deviance</li>
<li>The Akaike information criterion (AIC) is a metric for comparing different models to find the best fit
<ul>
<li>Notice the AIC score in your output?</li>
<li>This works the same way that it did in multiple linear regression</li>
</ul></li>
</ul>
</section>
<section id="what-is-deviance" class="level2">
<h2 class="anchored" data-anchor-id="what-is-deviance">What is Deviance?</h2>
<ul>
<li>Deviance measures model fit in <strong>generalized linear models (GLMs)</strong></li>
<li>Similar to residual sum of squares in linear regression</li>
<li>Compares fitted model to a <strong>saturated model</strong></li>
<li>Saturated model: fits data <strong>perfectly</strong> (one parameter per observation)</li>
</ul>
</section>
<section id="mathematical-definition" class="level2">
<h2 class="anchored" data-anchor-id="mathematical-definition">Mathematical Definition</h2>
<p><span class="math display">\[
\text{Deviance} = 2 \times (\log L_{\text{saturated}} - \log L_{\text{fitted}})
\]</span></p>
<ul>
<li>Lower deviance = better fit</li>
<li>Used to compare models</li>
</ul>
</section>
<section id="types-of-deviance-in-glm" class="level2">
<h2 class="anchored" data-anchor-id="types-of-deviance-in-glm">Types of Deviance in <code>glm()</code></h2>
<ul>
<li><strong>Null Deviance</strong>:
<ul>
<li>From model with <strong>only intercept</strong></li>
<li>Measures fit using just the <strong>mean response</strong></li>
</ul></li>
<li><strong>Residual Deviance</strong>:
<ul>
<li>From the <strong>fitted model</strong> (with predictors)</li>
<li>Measures how well your model fits relative to saturated model</li>
</ul></li>
</ul>
</section>
<section id="interpreting-deviance" class="level2">
<h2 class="anchored" data-anchor-id="interpreting-deviance">Interpreting Deviance</h2>
<ul>
<li>Large drop from null deviance to residual deviance → better model fit</li>
<li>Test improvement with chi-square test:</li>
</ul>
<div class="sourceCode" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">with</span>(model, <span class="fu">pchisq</span>(null.deviance <span class="sc">-</span> deviance, </span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>                   df.null <span class="sc">-</span> df.residual, </span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">lower.tail =</span> <span class="cn">FALSE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="example-output" class="level2">
<h2 class="anchored" data-anchor-id="example-output">Example Output</h2>
<div class="sourceCode" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>Null deviance<span class="sc">:</span> <span class="fl">120.45</span>  on <span class="dv">99</span>  degrees of freedom</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>Residual deviance<span class="sc">:</span>  <span class="fl">98.23</span>  on <span class="dv">97</span>  degrees of freedom</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>The model explains variability beyond just the mean</li>
<li>Compare deviances to assess predictor usefulness</li>
</ul>
</section>
</section>
<section id="generalized-linear-models-1" class="level1">
<h1>Generalized Linear Models</h1>
<section id="what-is-a-generalized-linear-model" class="level2">
<h2 class="anchored" data-anchor-id="what-is-a-generalized-linear-model">What is a generalized linear model?</h2>
<ul>
<li>The response variable of interest is binary - violates normality assumption because binomial distributed (<strong>Logistic</strong>)</li>
<li>The response variable is a count - violates the assumption of independent mean and variance because the data are multinomial or Poisson distributed (<strong>Poisson or Log-linear</strong>)</li>
<li>Provides a convenient way to connect predictor variables with categorical response variables</li>
<li>Basically an extension of general linear models that link the distribution of the response variable to the predictor variables</li>
</ul>
</section>
<section id="generalized-linear-models-have-three-components" class="level2">
<h2 class="anchored" data-anchor-id="generalized-linear-models-have-three-components">Generalized linear models have three components</h2>
<ul>
<li>The <strong>random component</strong> that specifies the residual distribution of the response variable in terms of a function of the mean.</li>
<li>The <strong>systematic component</strong> represents the linear combination of the predictor variables, and is equivalent to what you’ve learned previously for general linear models</li>
<li>The <strong>link function</strong> connects the random and systematic components, and depends on the nature of the random response distribution (e.g.&nbsp;binomial or poisson)</li>
</ul>
</section>
<section id="assumptions" class="level2">
<h2 class="anchored" data-anchor-id="assumptions">Assumptions</h2>
<ul>
<li>Independence of observations</li>
<li>No observations are overly influential</li>
<li>Linearity between the response and predictors is not assumed, but the relationship between each of the predictors and the link function is assumed to be linear</li>
<li>The dispersion is not extreme (over or under), and fits the assumptions for binomial or Poisson error distributions</li>
<li>What if the dispersion assumption is not met?
<ul>
<li>Can fit ‘quasibinomial’ or ‘quasipoisson’ distributions where the dispersion parameter is derived from the data</li>
<li>Use a Generalized Additive Model (GAM)</li>
</ul></li>
</ul>
</section>
</section>
<section id="logistic-regression" class="level1">
<h1>Logistic Regression</h1>
<section id="what-is-logistic-regression" class="level2">
<h2 class="anchored" data-anchor-id="what-is-logistic-regression">What is Logistic Regression?</h2>
<ul>
<li>A <strong>classification</strong> algorithm</li>
<li>Used to predict a <strong>binary outcome</strong> (e.g., success/failure)</li>
<li>Outputs a <strong>probability</strong> that is mapped to class labels</li>
</ul>
</section>
<section id="why-not-linear-regression" class="level2">
<h2 class="anchored" data-anchor-id="why-not-linear-regression">Why Not Linear Regression?</h2>
<ul>
<li>Linear regression is unbounded</li>
<li>Logistic regression uses the <strong>logit</strong> function to map outputs between 0 and 1</li>
</ul>
<p><span class="math display">\[P(y = 1 \mid x) = \frac{1}{1 + e^{-(\beta_0 + \beta_1 x)}}\]</span></p>
</section>
<section id="the-logistic-function" class="level2">
<h2 class="anchored" data-anchor-id="the-logistic-function">The Logistic Function</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">curve</span>(<span class="dv">1</span> <span class="sc">/</span> (<span class="dv">1</span> <span class="sc">+</span> <span class="fu">exp</span>(<span class="sc">-</span>x)), <span class="at">from =</span> <span class="sc">-</span><span class="dv">10</span>, <span class="at">to =</span> <span class="dv">10</span>, <span class="at">col =</span> <span class="st">"blue"</span>, <span class="at">ylab =</span> <span class="st">"Probability"</span>, <span class="at">xlab =</span> <span class="st">"Linear predictor"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Week8b_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>S-shaped (sigmoid) curve</li>
<li>Converts linear predictions into probabilities</li>
</ul>
</section>
<section id="odds-and-log-odds" class="level2">
<h2 class="anchored" data-anchor-id="odds-and-log-odds">Odds and Log-Odds</h2>
<p>Logistic regression models log-odds as a linear function</p>
<p><em><strong>Odds</strong>:</em> <span class="math display">\[\frac{P}{1 - P}\]</span></p>
<p><strong><em>Log-odds (logit)</em></strong>:</p>
<p><span class="math display">\[\log\left(\frac{P}{1 - P}\right)\]</span></p>
</section>
<section id="equivalent-model-equation" class="level2">
<h2 class="anchored" data-anchor-id="equivalent-model-equation">Equivalent Model Equation</h2>
<p><span class="math display">\[P(y = 1 \mid x) = \frac{1}{1 + e^{-(\beta_0 + \beta_1 x)}}\]</span></p>
<p><span class="math display">\[
\log\left(\frac{P(y=1)}{1 - P(y=1)}\right) = \beta_0 + \beta_1 x
\]</span></p>
<ul>
<li>Linear in parameters</li>
<li>Non-linear in terms of prediction</li>
<li>We have <strong>linked</strong> the two sides with the <code>logit</code> function</li>
</ul>
</section>
<section id="fitting-the-model" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="fitting-the-model">Fitting the Model</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Example in R</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(mtcars)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>mtcars<span class="sc">$</span>am <span class="ot">&lt;-</span> <span class="fu">factor</span>(mtcars<span class="sc">$</span>am)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">glm</span>(am <span class="sc">~</span> mpg, <span class="at">data =</span> mtcars, <span class="at">family =</span> binomial)</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = am ~ mpg, family = binomial, data = mtcars)

Coefficients:
            Estimate Std. Error z value Pr(&gt;|z|)   
(Intercept)  -6.6035     2.3514  -2.808  0.00498 **
mpg           0.3070     0.1148   2.673  0.00751 **
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 43.230  on 31  degrees of freedom
Residual deviance: 29.675  on 30  degrees of freedom
AIC: 33.675

Number of Fisher Scoring iterations: 5</code></pre>
</div>
</div>
<ul>
<li>Uses <strong>maximum likelihood estimation</strong></li>
<li>Output includes coefficients for log-odds</li>
</ul>
</section>
<section id="interpreting-coefficients" class="level2">
<h2 class="anchored" data-anchor-id="interpreting-coefficients">Interpreting Coefficients</h2>
<ul>
<li>Coefficients are in <strong>log-odds</strong></li>
<li>To interpret: exponentiate</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">exp</span>(<span class="fu">coef</span>(model))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>(Intercept)         mpg 
0.001355579 1.359379288 </code></pre>
</div>
</div>
<ul>
<li><span class="math inline">\(\exp(\beta_1)\)</span>: multiplicative change in odds for a 1-unit increase in predictor</li>
</ul>
</section>
<section id="making-predictions" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="making-predictions">Making Predictions</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">predict</span>(model, <span class="at">newdata =</span> <span class="fu">data.frame</span>(<span class="at">mpg =</span> <span class="dv">20</span>), <span class="at">type =</span> <span class="st">"response"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        1 
0.3862832 </code></pre>
</div>
</div>
<ul>
<li>Returns the <strong>probability</strong> that <code>am = 1</code> given <code>mpg = 20</code></li>
</ul>
</section>
<section id="model-evaluation-metrics" class="level2">
<h2 class="anchored" data-anchor-id="model-evaluation-metrics">Model Evaluation Metrics</h2>
<p>To assess the quality of a logistic regression model, we often use:</p>
<ul>
<li><strong>Confusion matrix</strong></li>
<li><strong>Accuracy</strong></li>
<li><strong>Precision</strong></li>
<li><strong>Recall</strong></li>
<li><strong>AUC-ROC</strong></li>
</ul>
</section>
<section id="confusion-matrix" class="level2">
<h2 class="anchored" data-anchor-id="confusion-matrix">Confusion Matrix</h2>
<p>A 2x2 table comparing predicted vs.&nbsp;actual values:</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th>Predicted Positive</th>
<th>Predicted Negative</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Actual Positive</strong></td>
<td>True Positive (TP)</td>
<td>False Negative (FN)</td>
</tr>
<tr class="even">
<td><strong>Actual Negative</strong></td>
<td>False Positive (FP)</td>
<td>True Negative (TN)</td>
</tr>
</tbody>
</table>
<p>Helps visualize performance across all classification outcomes.</p>
</section>
<section id="accuracy" class="level2">
<h2 class="anchored" data-anchor-id="accuracy">Accuracy</h2>
<ul>
<li>Proportion of <strong>correct predictions</strong> out of all predictions:</li>
</ul>
<p><span class="math display">\[
\text{Accuracy} = \frac{TP + TN}{TP + TN + FP + FN}
\]</span></p>
<ul>
<li>Simple but may be <strong>misleading</strong> with imbalanced data.</li>
</ul>
</section>
<section id="precision" class="level2">
<h2 class="anchored" data-anchor-id="precision">Precision</h2>
<ul>
<li>Of all predicted positives, how many are actually positive?</li>
</ul>
<p><span class="math display">\[
\text{Precision} = \frac{TP}{TP + FP}
\]</span></p>
<ul>
<li>High precision = <strong>low false positive rate</strong></li>
</ul>
</section>
<section id="recall-sensitivity" class="level2">
<h2 class="anchored" data-anchor-id="recall-sensitivity">Recall (Sensitivity)</h2>
<ul>
<li>Of all actual positives, how many were predicted correctly?</li>
</ul>
<p><span class="math display">\[
\text{Recall} = \frac{TP}{TP + FN}
\]</span></p>
<ul>
<li>High recall = <strong>low false negative rate</strong></li>
</ul>
</section>
<section id="precision-vs.-recall" class="level2">
<h2 class="anchored" data-anchor-id="precision-vs.-recall">Precision vs.&nbsp;Recall</h2>
<ul>
<li><strong>Trade-off</strong>: improving one can reduce the other</li>
<li>Use <strong>F1 score</strong> to balance both:</li>
</ul>
<p><span class="math display">\[
F1 = 2 \cdot \frac{\text{Precision} \cdot \text{Recall}}{\text{Precision} + \text{Recall}}
\]</span></p>
</section>
<section id="auc-roc" class="level2">
<h2 class="anchored" data-anchor-id="auc-roc">AUC-ROC</h2>
<ul>
<li><strong>ROC Curve</strong> plots TPR (Recall) vs.&nbsp;FPR across thresholds</li>
<li><strong>AUC</strong> (Area Under Curve) summarizes performance:</li>
<li>Probability the model ranks a random positive higher than a random negative</li>
<li>Ranges from <strong>0.5 (random)</strong> to <strong>1.0 (perfect)</strong></li>
</ul>
<hr>
</section>
<section id="r-interlude-logistic-regression" class="level2">
<h2 class="anchored" data-anchor-id="r-interlude-logistic-regression">R INTERLUDE | Logistic regression</h2>
</section>
<section id="dataset-variable-descriptions" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="dataset-variable-descriptions">Dataset Variable Descriptions</h2>
<table class="caption-top table">
<colgroup>
<col style="width: 36%">
<col style="width: 30%">
<col style="width: 32%">
</colgroup>
<thead>
<tr class="header">
<th>Variable</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>biomaterial</code></td>
<td>Categorical</td>
<td>Type of biomaterial used in the scaffold (<code>Material_A</code>, <code>Material_B</code>, <code>Material_C</code>)</td>
</tr>
<tr class="even">
<td><code>stiffness</code></td>
<td>Numeric</td>
<td>Mechanical stiffness of the scaffold (in kilopascals, kPa)</td>
</tr>
<tr class="odd">
<td><code>age_days</code></td>
<td>Integer</td>
<td>Age of the scaffold at implantation (in days; range: 20–60)</td>
</tr>
<tr class="even">
<td><code>success</code></td>
<td>Binary</td>
<td>Outcome of tissue integration (<code>1</code> = successful, <code>0</code> = unsuccessful)</td>
</tr>
</tbody>
</table>
</section>
<section id="fitting-the-model-1" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="fitting-the-model-1">Fitting the model</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>bio_data <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"BioE_logistic_data.csv"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">glm</span>(success <span class="sc">~</span> biomaterial <span class="sc">+</span> stiffness <span class="sc">+</span> age_days, </span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">data =</span> bio_data, </span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">family =</span> binomial)</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="fitting-the-model-2" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="fitting-the-model-2">Fitting the model</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = success ~ biomaterial + stiffness + age_days, family = binomial, 
    data = bio_data)

Coefficients:
                      Estimate Std. Error z value Pr(&gt;|z|)   
(Intercept)            1.72263    1.42408   1.210  0.22641   
biomaterialMaterial_B  0.84062    0.54300   1.548  0.12160   
biomaterialMaterial_C  1.34228    0.63184   2.124  0.03364 * 
stiffness              0.23744    0.11396   2.084  0.03720 * 
age_days              -0.06895    0.02441  -2.825  0.00473 **
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 135.17  on 149  degrees of freedom
Residual deviance: 114.89  on 145  degrees of freedom
AIC: 124.89

Number of Fisher Scoring iterations: 5</code></pre>
</div>
</div>
</section>
<section id="plotting-the-model" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="plotting-the-model">Plotting the model</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bio_data, <span class="fu">aes</span>(<span class="at">x =</span> stiffness, <span class="at">y =</span> success, <span class="at">color =</span> biomaterial)) <span class="sc">+</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.4</span>, <span class="at">shape =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"glm"</span>, <span class="at">method.args =</span> <span class="fu">list</span>(<span class="at">family =</span> <span class="st">"binomial"</span>), </span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">formula =</span> success <span class="sc">~</span> biomaterial <span class="sc">+</span> stiffness <span class="sc">+</span> age_days, <span class="at">se =</span> <span class="cn">FALSE</span>, <span class="at">size =</span> <span class="fl">1.2</span>) <span class="sc">+</span></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Predicted Probability of Success"</span>,</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Stiffness (kPa)"</span>,</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Predicted Probability"</span>,</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">color =</span> <span class="st">"Biomaterial"</span>) <span class="sc">+</span></span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
ℹ Please use `linewidth` instead.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Failed to fit group 1.
Caused by error:
! object 'success' not found</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Failed to fit group 2.
Caused by error:
! object 'success' not found</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Failed to fit group 3.
Caused by error:
! object 'success' not found</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Week8b_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
</section>
<section id="poisson-regression" class="level1">
<h1>Poisson Regression</h1>
<section id="what-is-poisson-regression" class="level2">
<h2 class="anchored" data-anchor-id="what-is-poisson-regression">What is Poisson Regression?</h2>
<ul>
<li>A type of <strong>Generalized Linear Model (GLM)</strong></li>
<li>Used to model <strong>count data</strong> (e.g., number of visits, accidents, events)</li>
<li>The outcome variable is assumed to follow a <strong>Poisson distribution</strong></li>
</ul>
</section>
<section id="the-poisson-distribution" class="level2">
<h2 class="anchored" data-anchor-id="the-poisson-distribution">The Poisson Distribution</h2>
<p><span class="math display">\[ P(Y = y) = \frac{e^{-\lambda} \lambda^y}{y!} \]</span></p>
<ul>
<li>Mean = Variance = <span class="math inline">\(\lambda\)</span></li>
<li>Suitable for modeling counts of <strong>rare events</strong></li>
</ul>
</section>
<section id="model-structure" class="level2">
<h2 class="anchored" data-anchor-id="model-structure">Model Structure</h2>
<p><span class="math display">\[ Y_i \sim \text{Poisson}(\lambda_i) \]</span></p>
<p><span class="math display">\[
\log(\lambda_i) = \beta_0 + \beta_1 x_{i1} + \ldots + \beta_p x_{ip}
\]</span></p>
<ul>
<li>The <strong>log link</strong> ensures ( _i &gt; 0 )</li>
<li>Linear model for the <strong>log of expected counts</strong></li>
</ul>
</section>
<section id="why-not-use-linear-regression" class="level2">
<h2 class="anchored" data-anchor-id="why-not-use-linear-regression">Why Not Use Linear Regression?</h2>
<ul>
<li>Linear regression may predict <strong>negative counts</strong></li>
<li>It assumes <strong>constant variance</strong>, but Poisson variance = mean</li>
<li>Poisson regression accounts for the distributional structure of count data</li>
</ul>
</section>
<section id="example" class="level2">
<h2 class="anchored" data-anchor-id="example">Example</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">glm</span>(y <span class="sc">~</span> x1 <span class="sc">+</span> x2, <span class="at">data =</span> something, <span class="at">family =</span> poisson)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="interpreting-coefficients-1" class="level2">
<h2 class="anchored" data-anchor-id="interpreting-coefficients-1">Interpreting Coefficients</h2>
<ul>
<li>Coefficients are on the <strong>log scale</strong></li>
<li>To interpret, <strong>exponentiate</strong> them:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">exp</span>(<span class="fu">coef</span>(model))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li><span class="math inline">\(\exp(\beta_j)\)</span>: multiplicative change in expected count for a 1-unit increase in predictor <span class="math inline">\(x_j\)</span></li>
</ul>
</section>
<section id="model-assumptions" class="level2">
<h2 class="anchored" data-anchor-id="model-assumptions">Model Assumptions</h2>
<ol type="1">
<li><strong>Counts are Poisson-distributed</strong></li>
<li><strong>Mean equals variance</strong> (can be relaxed in quasi-Poisson or negative binomial models)</li>
<li><strong>Log-linear relationship</strong> between predictors and the outcome</li>
</ol>
</section>
<section id="model-evaluation" class="level2">
<h2 class="anchored" data-anchor-id="model-evaluation">Model Evaluation</h2>
<ul>
<li><strong>Deviance residuals</strong></li>
<li><strong>Akaike Information Criterion (AIC)</strong></li>
<li><strong>Overdispersion checks</strong></li>
<li><strong>Pseudo-</strong><span class="math inline">\(R^2\)</span></li>
</ul>
</section>
<section id="overdispersion" class="level2">
<h2 class="anchored" data-anchor-id="overdispersion">Overdispersion</h2>
<ul>
<li>When variance &gt; mean</li>
<li>Can inflate type I errors</li>
<li>Check with:</li>
</ul>
<p><span class="math display">\[
\text{Dispersion} = \frac{\text{Residual deviance}}{\text{df}}
\]</span></p>
<ul>
<li><p>Consider alternatives:</p>
<ul>
<li><strong>Quasi-Poisson</strong></li>
<li><strong>Negative Binomial Regression</strong></li>
</ul></li>
</ul>
</section>
<section id="goodness-of-fit" class="level2">
<h2 class="anchored" data-anchor-id="goodness-of-fit">Goodness-of-Fit</h2>
<ul>
<li><strong>Deviance</strong>: compares model to saturated model</li>
<li><strong>Pearson residuals</strong>: detect outliers and overdispersion</li>
<li>Use <code>AIC</code> to compare models</li>
</ul>
</section>
<section id="r-interlude-log-linear-poisson-regression" class="level2">
<h2 class="anchored" data-anchor-id="r-interlude-log-linear-poisson-regression">R INTERLUDE | Log-linear (Poisson) regression</h2>
</section>
<section id="dataset-variable-descriptions-poisson-model" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="dataset-variable-descriptions-poisson-model">Dataset Variable Descriptions (Poisson Model)</h2>
<table class="caption-top table">
<colgroup>
<col style="width: 44%">
<col style="width: 26%">
<col style="width: 29%">
</colgroup>
<thead>
<tr class="header">
<th>Variable</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>biomaterial</code></td>
<td>Categorical</td>
<td>Type of biomaterial used in the scaffold</td>
</tr>
<tr class="even">
<td><code>stiffness</code></td>
<td>Numeric</td>
<td>Mechanical stiffness in kilopascals (kPa)</td>
</tr>
<tr class="odd">
<td><code>age_days</code></td>
<td>Integer</td>
<td>Age of scaffold at implantation (20–60 days)</td>
</tr>
<tr class="even">
<td><code>integration_events</code></td>
<td>Count</td>
<td>Number of successful tissue integration events (simulated)</td>
</tr>
</tbody>
</table>
</section>
<section id="fit-the-model" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="fit-the-model">Fit the model</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load data</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>bio_data <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"BioE_poisson_data.csv"</span>)</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit log-linear (Poisson) model</span></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">glm</span>(integration_events <span class="sc">~</span> biomaterial <span class="sc">+</span> stiffness <span class="sc">+</span> age_days,</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">family =</span> poisson, <span class="at">data =</span> bio_data)</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="fit-the-model-1" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="fit-the-model-1">Fit the model</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = integration_events ~ biomaterial + stiffness + 
    age_days, family = poisson, data = bio_data)

Coefficients:
                      Estimate Std. Error z value Pr(&gt;|z|)    
(Intercept)            1.11533    0.16566   6.733 1.67e-11 ***
biomaterialMaterial_B  0.42231    0.07179   5.883 4.03e-09 ***
biomaterialMaterial_C  0.67282    0.06835   9.843  &lt; 2e-16 ***
stiffness              0.14544    0.01226  11.864  &lt; 2e-16 ***
age_days              -0.01885    0.00221  -8.533  &lt; 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for poisson family taken to be 1)

    Null deviance: 465.13  on 149  degrees of freedom
Residual deviance: 151.52  on 145  degrees of freedom
AIC: 755.3

Number of Fisher Scoring iterations: 4</code></pre>
</div>
</div>
</section>
<section id="make-a-plot" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="make-a-plot">Make a plot</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Add predicted counts</span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>bio_data<span class="sc">$</span>predicted_counts <span class="ot">&lt;-</span> <span class="fu">predict</span>(model, <span class="at">type =</span> <span class="st">"response"</span>)</span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot predicted vs actual across stiffness for each biomaterial</span></span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bio_data, <span class="fu">aes</span>(<span class="at">x =</span> stiffness, <span class="at">y =</span> predicted_counts, <span class="at">color =</span> biomaterial)) <span class="sc">+</span></span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">y =</span> integration_events), <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">shape =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">size =</span> <span class="fl">1.2</span>) <span class="sc">+</span></span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Predicted Integration Events"</span>,</span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Stiffness (kPa)"</span>,</span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Predicted Count"</span>,</span>
<span id="cb44-14"><a href="#cb44-14" aria-hidden="true" tabindex="-1"></a>       <span class="at">color =</span> <span class="st">"Biomaterial"</span>) <span class="sc">+</span></span>
<span id="cb44-15"><a href="#cb44-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
</section>
<section id="section" class="level1">
<h1></h1>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>