<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.30">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Bill Cresko">

<title>Week 6a - Statistics for Bioengineering – BioEngineering Statistics Winter 2026</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-de070a7b0ab54f8780927367ac907214.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-f7fe251563e67c05764dc4775ff48cfc.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">BioEngineering Statistics Winter 2026</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../Syllabus.html"> 
<span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../Policies.html"> 
<span class="menu-text">Policies</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../Schedule.html"> 
<span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../Lectures.html"> 
<span class="menu-text">Lectures</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../Homework.html"> 
<span class="menu-text">Homeworks</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../Software.html"> 
<span class="menu-text">Software</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../Resources.html"> 
<span class="menu-text">Resources</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#this-week" id="toc-this-week" class="nav-link active" data-scroll-target="#this-week">This week</a></li>
  <li><a href="#linear-models-and-regression" id="toc-linear-models-and-regression" class="nav-link" data-scroll-target="#linear-models-and-regression">Linear Models and Regression</a>
  <ul class="collapse">
  <li><a href="#general-linear-model---two-or-more-continuous-variables" id="toc-general-linear-model---two-or-more-continuous-variables" class="nav-link" data-scroll-target="#general-linear-model---two-or-more-continuous-variables">General linear model - two or more continuous variables</a></li>
  <li><a href="#generalized-linear-models---non-linear-in-the-coefficients" id="toc-generalized-linear-models---non-linear-in-the-coefficients" class="nav-link" data-scroll-target="#generalized-linear-models---non-linear-in-the-coefficients">Generalized linear models - non-linear in the coefficients</a></li>
  </ul></li>
  <li><a href="#model-1-and-model-2-regression" id="toc-model-1-and-model-2-regression" class="nav-link" data-scroll-target="#model-1-and-model-2-regression">Model 1 and Model 2 regression</a>
  <ul class="collapse">
  <li><a href="#hypothesis-tests-in-regression" id="toc-hypothesis-tests-in-regression" class="nav-link" data-scroll-target="#hypothesis-tests-in-regression">Hypothesis tests in regression</a></li>
  <li><a href="#section" id="toc-section" class="nav-link" data-scroll-target="#section"></a></li>
  <li><a href="#section-1" id="toc-section-1" class="nav-link" data-scroll-target="#section-1"></a></li>
  <li><a href="#section-2" id="toc-section-2" class="nav-link" data-scroll-target="#section-2"></a></li>
  <li><a href="#r-interlude-model-i-and-ii-regression" id="toc-r-interlude-model-i-and-ii-regression" class="nav-link" data-scroll-target="#r-interlude-model-i-and-ii-regression">R INTERLUDE | Model I and II regression</a></li>
  <li><a href="#r-interlude-model-i-and-ii-regression-1" id="toc-r-interlude-model-i-and-ii-regression-1" class="nav-link" data-scroll-target="#r-interlude-model-i-and-ii-regression-1">R INTERLUDE | Model I and II regression</a></li>
  <li><a href="#r-interlude-model-i-and-ii-regression-2" id="toc-r-interlude-model-i-and-ii-regression-2" class="nav-link" data-scroll-target="#r-interlude-model-i-and-ii-regression-2">R INTERLUDE | Model I and II regression</a></li>
  <li><a href="#smoothers-and-local-regression" id="toc-smoothers-and-local-regression" class="nav-link" data-scroll-target="#smoothers-and-local-regression">Smoothers and local regression</a></li>
  <li><a href="#r-interlude-model-i-and-ii-regression-3" id="toc-r-interlude-model-i-and-ii-regression-3" class="nav-link" data-scroll-target="#r-interlude-model-i-and-ii-regression-3">R INTERLUDE | Model I and II regression</a></li>
  <li><a href="#r-interlude---on-your-own" id="toc-r-interlude---on-your-own" class="nav-link" data-scroll-target="#r-interlude---on-your-own">R INTERLUDE - on your own</a></li>
  </ul></li>
  <li><a href="#mutliple-linear-regression" id="toc-mutliple-linear-regression" class="nav-link" data-scroll-target="#mutliple-linear-regression">Mutliple Linear regression</a>
  <ul class="collapse">
  <li><a href="#multiple-linear-regression---goals" id="toc-multiple-linear-regression---goals" class="nav-link" data-scroll-target="#multiple-linear-regression---goals">Multiple Linear Regression - Goals</a></li>
  <li><a href="#multiple-linear-regression-additive-and-multiplicative-models-of-2-or-more-predictors" id="toc-multiple-linear-regression-additive-and-multiplicative-models-of-2-or-more-predictors" class="nav-link" data-scroll-target="#multiple-linear-regression-additive-and-multiplicative-models-of-2-or-more-predictors">Multiple Linear Regression | Additive and multiplicative models of 2 or more predictors</a></li>
  <li><a href="#multiple-linear-regression-additive-and-multiplicative-models" id="toc-multiple-linear-regression-additive-and-multiplicative-models" class="nav-link" data-scroll-target="#multiple-linear-regression-additive-and-multiplicative-models">Multiple Linear Regression | Additive and multiplicative models</a></li>
  <li><a href="#multiple-linear-regression-assumptions" id="toc-multiple-linear-regression-assumptions" class="nav-link" data-scroll-target="#multiple-linear-regression-assumptions">Multiple linear regression assumptions</a></li>
  <li><a href="#checking-for-multi-collinearity" id="toc-checking-for-multi-collinearity" class="nav-link" data-scroll-target="#checking-for-multi-collinearity">checking for multi-collinearity</a></li>
  <li><a href="#r-interlude-multiple-regression-with-2-predictor-variables" id="toc-r-interlude-multiple-regression-with-2-predictor-variables" class="nav-link" data-scroll-target="#r-interlude-multiple-regression-with-2-predictor-variables">R INTERLUDE | multiple regression with 2 predictor variables</a></li>
  <li><a href="#r-interlude-multiple-regression-with-2-predictor-variables-1" id="toc-r-interlude-multiple-regression-with-2-predictor-variables-1" class="nav-link" data-scroll-target="#r-interlude-multiple-regression-with-2-predictor-variables-1">R INTERLUDE | multiple regression with 2 predictor variables</a></li>
  <li><a href="#r-interlude-multiple-regression-with-2-predictor-variables-2" id="toc-r-interlude-multiple-regression-with-2-predictor-variables-2" class="nav-link" data-scroll-target="#r-interlude-multiple-regression-with-2-predictor-variables-2">R INTERLUDE | multiple regression with 2 predictor variables</a></li>
  <li><a href="#r-interlude-adding-a-polynomial" id="toc-r-interlude-adding-a-polynomial" class="nav-link" data-scroll-target="#r-interlude-adding-a-polynomial">R INTERLUDE | adding a polynomial</a></li>
  <li><a href="#r-interlude-adding-a-polynomial-1" id="toc-r-interlude-adding-a-polynomial-1" class="nav-link" data-scroll-target="#r-interlude-adding-a-polynomial-1">R INTERLUDE | adding a polynomial</a></li>
  <li><a href="#r-interlude-adding-a-polynomial-2" id="toc-r-interlude-adding-a-polynomial-2" class="nav-link" data-scroll-target="#r-interlude-adding-a-polynomial-2">R INTERLUDE | adding a polynomial</a></li>
  <li><a href="#r-interlude-adding-a-polynomial-3" id="toc-r-interlude-adding-a-polynomial-3" class="nav-link" data-scroll-target="#r-interlude-adding-a-polynomial-3">R INTERLUDE | adding a polynomial</a></li>
  </ul></li>
  <li><a href="#model-selection-when-you-have-many-predictor-variables" id="toc-model-selection-when-you-have-many-predictor-variables" class="nav-link" data-scroll-target="#model-selection-when-you-have-many-predictor-variables">Model selection when you have many predictor variables</a>
  <ul class="collapse">
  <li><a href="#model-selection" id="toc-model-selection" class="nav-link" data-scroll-target="#model-selection">Model selection</a></li>
  <li><a href="#model-selection-the-problems" id="toc-model-selection-the-problems" class="nav-link" data-scroll-target="#model-selection-the-problems">Model selection | the problems</a></li>
  <li><a href="#model-selection-a-beetle-example" id="toc-model-selection-a-beetle-example" class="nav-link" data-scroll-target="#model-selection-a-beetle-example">Model selection | a beetle example</a></li>
  <li><a href="#start-with-linear-regression" id="toc-start-with-linear-regression" class="nav-link" data-scroll-target="#start-with-linear-regression">Start with linear regression</a></li>
  <li><a href="#quadratic-2nd-degree-polynomial" id="toc-quadratic-2nd-degree-polynomial" class="nav-link" data-scroll-target="#quadratic-2nd-degree-polynomial">Quadratic (2nd degree) polynomial?</a></li>
  <li><a href="#quadratic-3rd-degree-polynomial" id="toc-quadratic-3rd-degree-polynomial" class="nav-link" data-scroll-target="#quadratic-3rd-degree-polynomial">Quadratic (3rd degree) polynomial?</a></li>
  <li><a href="#a-polynomial-of-degree-5" id="toc-a-polynomial-of-degree-5" class="nav-link" data-scroll-target="#a-polynomial-of-degree-5">A polynomial of degree 5?</a></li>
  <li><a href="#a-polynomial-of-degree-10" id="toc-a-polynomial-of-degree-10" class="nav-link" data-scroll-target="#a-polynomial-of-degree-10">A polynomial of degree 10?</a></li>
  <li><a href="#the-problem-with-this-approach" id="toc-the-problem-with-this-approach" class="nav-link" data-scroll-target="#the-problem-with-this-approach">The problem with this approach</a></li>
  <li><a href="#the-problem-with-this-approach-1" id="toc-the-problem-with-this-approach-1" class="nav-link" data-scroll-target="#the-problem-with-this-approach-1">The problem with this approach</a></li>
  <li><a href="#lets-consider-our-objectives" id="toc-lets-consider-our-objectives" class="nav-link" data-scroll-target="#lets-consider-our-objectives">Let’s consider our objectives</a></li>
  <li><a href="#a-polynomial-of-degree-10-1" id="toc-a-polynomial-of-degree-10-1" class="nav-link" data-scroll-target="#a-polynomial-of-degree-10-1">A polynomial of degree 10?</a></li>
  <li><a href="#section-3" id="toc-section-3" class="nav-link" data-scroll-target="#section-3"></a></li>
  </ul></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="Week6a.html"><i class="bi bi-file-slides"></i>RevealJS</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Week 6a - Statistics for Bioengineering</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Bill Cresko </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'ggplot2' was built under R version 4.4.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'tidyr' was built under R version 4.4.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'readr' was built under R version 4.4.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.6
✔ forcats   1.0.1     ✔ stringr   1.6.0
✔ ggplot2   4.0.1     ✔ tibble    3.3.0
✔ lubridate 1.9.4     ✔ tidyr     1.3.2
✔ purrr     1.2.0     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'gt' was built under R version 4.4.3</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="this-week" class="level2">
<h2 class="anchored" data-anchor-id="this-week">This week</h2>
<ul>
<li><p>Model 1 and 2 regression</p></li>
<li><p>Multiple linear regression</p></li>
<li><p>Analysis of Variance (ANOVA)</p></li>
<li><p>Key principles of experimental design</p></li>
<li><p>How to report statistics in papers</p></li>
</ul>
</section>
<section id="linear-models-and-regression" class="level1">
<h1>Linear Models and Regression</h1>
<section id="general-linear-model---two-or-more-continuous-variables" class="level2">
<h2 class="anchored" data-anchor-id="general-linear-model---two-or-more-continuous-variables">General linear model - two or more continuous variables</h2>
<p><span class="math display">\[ y_i = \beta_0 + \beta_1 x_1 + \epsilon_i \]</span></p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/images_4b.008.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
</figure>
</div>
</div>
</div>
<p><span class="math display">\[ y_i = \beta_0 + \beta_1 x_1 + \beta_2 x_2 + ... + \beta_n x_n + \epsilon_i \]</span></p>
<p><span class="math display">\[ y_i = \beta_0 + \beta_1 x_1 + \beta_2 x_2 + \beta_3 x_1x_2 + \epsilon_i \]</span></p>
</section>
<section id="generalized-linear-models---non-linear-in-the-coefficients" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="generalized-linear-models---non-linear-in-the-coefficients">Generalized linear models - non-linear in the coefficients</h2>
<ul>
<li><p><strong>Polynomial regression</strong> <span class="math display">\[y = \beta_0 + \beta_1 x + \beta_2 x^2 +...+ \beta_n x^n +\epsilon\]</span></p></li>
<li><p><strong>Logistic regression</strong> <span class="math display">\[log(\frac{\pi_i}{1-\pi_1})=\beta_0 + \beta_1 x_1 + \beta_2 x_2 +...+ \beta_n x_n +\epsilon\]</span></p></li>
<li><p><strong>Poisson regression</strong> <span class="math display">\[log(y) = \beta_0 + \beta_1 x + \epsilon\]</span></p></li>
<li><p><strong>Exponential regression</strong> <span class="math display">\[y = \alpha^{\beta x} + \epsilon\]</span></p></li>
<li><p><strong>Power regression</strong> <span class="math display">\[y = \alpha x^\beta + \epsilon\]</span></p></li>
</ul>
</section>
</section>
<section id="model-1-and-model-2-regression" class="level1">
<h1>Model 1 and Model 2 regression</h1>
<section id="hypothesis-tests-in-regression" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="hypothesis-tests-in-regression">Hypothesis tests in regression</h2>
<p><span class="math display">\[H_0 : \beta_0 = 0\]</span> <span class="math display">\[H_A : \beta_1 \neq 0\]</span></p>
<p>reduced model (<span class="math inline">\(H_0\)</span>) -</p>
<p><span class="math display">\[y_i = \beta_0 + 0x_i + \epsilon_i\]</span></p>
<p>full model (<span class="math inline">\(H_A\)</span>) -</p>
<p><span class="math display">\[y_i = \beta_0 + \beta_1x_i + \epsilon_i\]</span><br>
<br>
</p>
<ol type="1">
<li>fits a “reduced” model without slope term (H0)</li>
<li>fits the “full” model with slope term added back</li>
<li>compares fit of full and reduced models using an <strong><em>F test statistic</em></strong></li>
<li>compares the residual variance of the full vs.&nbsp;reduced models</li>
</ol>
</section>
<section id="section" class="level2">
<h2 class="anchored" data-anchor-id="section"></h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/images_5a.001.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="section-1" class="level2">
<h2 class="anchored" data-anchor-id="section-1"></h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/images_5a.002.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="section-2" class="level2">
<h2 class="anchored" data-anchor-id="section-2"></h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/images_5a.003.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
<section id="r-interlude-model-i-and-ii-regression" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="r-interlude-model-i-and-ii-regression">R INTERLUDE | Model I and II regression</h2>
<ul>
<li>First we fit a standard model I ordinary least squares</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>zfish_data <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Drerio_development_complete.csv"</span>, <span class="at">header=</span>T, <span class="at">sep=</span><span class="st">","</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>length <span class="ot">&lt;-</span> zfish_data<span class="sc">$</span>Length_cm</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>weight <span class="ot">&lt;-</span> zfish_data<span class="sc">$</span>Weight_mg</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>size_lm <span class="ot">&lt;-</span> <span class="fu">lm</span>(weight<span class="sc">~</span>length)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(size_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = weight ~ length)

Residuals:
    Min      1Q  Median      3Q     Max 
-4.1164 -1.1004  0.0976  0.9950  4.5447 

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)  2.42136    0.12820   18.89   &lt;2e-16 ***
length       3.06303    0.03775   81.13   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 1.48 on 598 degrees of freedom
Multiple R-squared:  0.9167,    Adjusted R-squared:  0.9166 
F-statistic:  6582 on 1 and 598 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
</section>
<section id="r-interlude-model-i-and-ii-regression-1" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="r-interlude-model-i-and-ii-regression-1">R INTERLUDE | Model I and II regression</h2>
<ul>
<li>We can fit the line</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(size_lm, <span class="fu">aes</span>(<span class="at">x =</span> weight, <span class="at">y =</span> length)) <span class="sc">+</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">color =</span> <span class="st">"blue"</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span>                  </span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>, <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">se =</span> <span class="cn">TRUE</span>) <span class="sc">+</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"Linear regression of zebrafish weight on standard length"</span>,</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"zebrafish length in cm"</span>,</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">"zebrafish weight in gm"</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: `fortify(&lt;lm&gt;)` was deprecated in ggplot2 4.0.0.
ℹ Please use `broom::augment(&lt;lm&gt;)` instead.
ℹ The deprecated feature was likely used in the ggplot2 package.
  Please report the issue at &lt;https://github.com/tidyverse/ggplot2/issues&gt;.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Week6a_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="r-interlude-model-i-and-ii-regression-2" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="r-interlude-model-i-and-ii-regression-2">R INTERLUDE | Model I and II regression</h2>
<ul>
<li>We can fit a <code>LOESS</code> line</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(size_lm, <span class="fu">aes</span>(<span class="at">x =</span> weight, <span class="at">y =</span> length)) <span class="sc">+</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">color =</span> <span class="st">"blue"</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span>                  </span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"loess"</span>, <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">se =</span> <span class="cn">TRUE</span>) <span class="sc">+</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"Linear regression of zebrafish weight on standard length"</span>,</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"zebrafish length in cm"</span>,</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">"zebrafish weight in gm"</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Week6a_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="smoothers-and-local-regression" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="smoothers-and-local-regression">Smoothers and local regression</h2>
<ul>
<li><strong>loess</strong> (locally estimated scatterplot smoothing) - fit least squares regression lines to successive subsets of the observations weighted according to their distance from the focal observation.<br>
</li>
<li><strong>kernel smoothers</strong> - new smoothed y-values are computed as the weighted averages of points within a defined window of the original x-values. Weightings are determined by the type of kernel smoother specified, the larger the window, the greater the degree of smoothing.<br>
</li>
<li><strong>splines</strong> - joins together a series of polynomial fits that have been generated after the data is split up into a number of smaller windows.</li>
</ul>
</section>
<section id="r-interlude-model-i-and-ii-regression-3" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="r-interlude-model-i-and-ii-regression-3">R INTERLUDE | Model I and II regression</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(lmodel2)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>size_lm_ModII <span class="ot">&lt;-</span> <span class="fu">lmodel2</span>(weight <span class="sc">~</span> length)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>RMA was not requested: it will not be computed.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>No permutation test will be performed</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>size_lm_ModII</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Model II regression

Call: lmodel2(formula = weight ~ length)

n = 600   r = 0.9574514   r-square = 0.9167132 
Parametric P-values:   2-tailed = 0    1-tailed = 0 
Angle between the two OLS regression lines = 1.418965 degrees

Regression results
  Method Intercept    Slope Angle (degrees) P-perm (1-tailed)
1    OLS  2.421359 3.063034        71.91950                NA
2     MA  1.662671 3.316382        73.22019                NA
3    SMA  2.013729 3.199153        72.64166                NA

Confidence intervals
  Method 2.5%-Intercept 97.5%-Intercept 2.5%-Slope 97.5%-Slope
1    OLS       2.169579        2.673139   2.988886    3.137182
2     MA       1.416796        1.897871   3.237842    3.398487
3    SMA       1.789108        2.233204   3.125864    3.274161

Eigenvalues: 28.63954 0.1960763 

H statistic used for computing C.I. of MA: 4.476936e-05 </code></pre>
</div>
</div>
<ul>
<li>You should also have generated <code>OLS (Ordinary Least Squares)</code>, <code>MA (Moving Average)</code>, and <code>SMA (Standard Major Axis)</code> regression models, and the last plots SMA line from parameter estimates</li>
</ul>
</section>
<section id="r-interlude---on-your-own" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="r-interlude---on-your-own">R INTERLUDE - on your own</h2>
<ul>
<li><p>Say you measured green fluorescent protein abundance in cell culture across several key, carefully controlled temperatures.</p></li>
<li><p>You ultimately want to know whether protein expression changes as a function of temperature in your experiment.</p></li>
<li><p>Read in the <code>gfp_temp.tsv</code> data and perform a regression analysis.</p></li>
<li><p>Address the following questions</p>
<ul>
<li>Which is X and which is Y?</li>
<li>Are residual assumptions met?</li>
<li>What model of regression should we use in this case and why?</li>
<li>What is our decision regarding our null hypothesis of no relationship?</li>
</ul></li>
</ul>
</section>
</section>
<section id="mutliple-linear-regression" class="level1">
<h1>Mutliple Linear regression</h1>
<section id="multiple-linear-regression---goals" class="level2">
<h2 class="anchored" data-anchor-id="multiple-linear-regression---goals">Multiple Linear Regression - Goals</h2>
<ul>
<li><p>To develop a better predictive model than is possible from models based on single independent variables.</p></li>
<li><p>To investigate the relative individual effects of each of the multiple independent variables above and beyond the effects of the other variables.</p></li>
<li><p>The individual effects of each of the predictor variables on the response variable can be depicted by single partial regression lines.</p></li>
<li><p>The slope of any single partial regression line (partial regression slope) thereby represents the rate of change or effect of that specific predictor variable (holding all the other predictor variables constant to their respective mean values) on the response variable.</p></li>
</ul>
</section>
<section id="multiple-linear-regression-additive-and-multiplicative-models-of-2-or-more-predictors" class="level2">
<h2 class="anchored" data-anchor-id="multiple-linear-regression-additive-and-multiplicative-models-of-2-or-more-predictors">Multiple Linear Regression | Additive and multiplicative models of 2 or more predictors</h2>
<p>Additive model <span class="math display">\[y_i = \beta_0 + \beta_1x_{1} + \beta_2x_{2} + ... + \beta_jx_{j} + \epsilon_i\]</span></p>
<p><br>
</p>
<p>Multiplicative model (with two predictors) <span class="math display">\[y_i = \beta_0 + \beta_1x_1 + \beta_2x_2 + \beta_3x_1x_2 + \epsilon_i\]</span></p>
</section>
<section id="multiple-linear-regression-additive-and-multiplicative-models" class="level2">
<h2 class="anchored" data-anchor-id="multiple-linear-regression-additive-and-multiplicative-models">Multiple Linear Regression | Additive and multiplicative models</h2>
<p><br></p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/images_5a.006.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="multiple-linear-regression-assumptions" class="level2">
<h2 class="anchored" data-anchor-id="multiple-linear-regression-assumptions">Multiple linear regression assumptions</h2>
<p><br></p>
<ul>
<li>linearity</li>
<li>normality</li>
<li>homogeneity of variance</li>
<li>The absence of strong <strong>multi-collinearity</strong>
<ul>
<li>a predictor variable must not be linearly predicted by - or correlated with - any combination of the other predictor variables.</li>
</ul></li>
</ul>
</section>
<section id="checking-for-multi-collinearity" class="level2">
<h2 class="anchored" data-anchor-id="checking-for-multi-collinearity">checking for multi-collinearity</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(car)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: carData</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'car'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:dplyr':

    recode</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:purrr':

    some</code></pre>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>zfish_data <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Drerio_development_complete.csv"</span>, <span class="at">header=</span>T, <span class="at">sep=</span><span class="st">","</span>)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>length <span class="ot">&lt;-</span> zfish_data<span class="sc">$</span>Length_cm</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>weight <span class="ot">&lt;-</span> zfish_data<span class="sc">$</span>Weight_mg</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>age <span class="ot">&lt;-</span> zfish_data<span class="sc">$</span>Age_Days</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>pigment <span class="ot">&lt;-</span> zfish_data<span class="sc">$</span>Pigmentation</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a><span class="fu">scatterplotMatrix</span>(<span class="sc">~</span>length<span class="sc">+</span>pigment<span class="sc">+</span>age, <span class="at">diag=</span><span class="st">"boxplot"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in applyDefaults(diagonal, defaults = list(method = "adaptiveDensity"),
: unnamed diag arguments, will be ignored</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Week6a_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="r-interlude-multiple-regression-with-2-predictor-variables" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="r-interlude-multiple-regression-with-2-predictor-variables">R INTERLUDE | multiple regression with 2 predictor variables</h2>
<ul>
<li>Read in the <code>RNAseq_lipid.tsv</code> and examine the continuous variables.</li>
<li>We are interested in whether Gene01 and/or Gene02 expression levels influence lipid content.</li>
<li>First plot <span class="math inline">\(Y\)</span> vs <span class="math inline">\(Gene01\)</span>, <span class="math inline">\(Y\)</span> vs <span class="math inline">\(Gene02\)</span>, then set up and test a multiplicative model.</li>
<li>What are the parameter estimates of interest in this case?</li>
<li>What are the outcomes from our hypothesis tests?</li>
</ul>
</section>
<section id="r-interlude-multiple-regression-with-2-predictor-variables-1" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="r-interlude-multiple-regression-with-2-predictor-variables-1">R INTERLUDE | multiple regression with 2 predictor variables</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>RNAseq_Data <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"RNAseq_lipid.tsv"</span>, <span class="at">header=</span>T, <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> RNAseq_Data<span class="sc">$</span>Lipid_Conc</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> RNAseq_Data<span class="sc">$</span>Gene01</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>g2 <span class="ot">&lt;-</span> RNAseq_Data<span class="sc">$</span>Gene02</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>g3 <span class="ot">&lt;-</span> RNAseq_Data<span class="sc">$</span>Gene03</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>g4 <span class="ot">&lt;-</span> RNAseq_Data<span class="sc">$</span>Gene04</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>Mult_lm <span class="ot">&lt;-</span> <span class="fu">lm</span>(y <span class="sc">~</span> g1<span class="sc">*</span>g2)</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(Mult_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = y ~ g1 * g2)

Residuals:
     Min       1Q   Median       3Q      Max 
-0.52111 -0.16333 -0.00931  0.21515  0.54255 

Coefficients:
              Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)  3.2965521  0.3461603   9.523 5.40e-08 ***
g1           0.0558232  0.0040117  13.915 2.34e-10 ***
g2          -0.0082765  0.0290978  -0.284    0.780    
g1:g2        0.0001702  0.0003199   0.532    0.602    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.2997 on 16 degrees of freedom
Multiple R-squared:  0.9846,    Adjusted R-squared:  0.9817 
F-statistic: 341.3 on 3 and 16 DF,  p-value: 1.042e-14</code></pre>
</div>
</div>
</section>
<section id="r-interlude-multiple-regression-with-2-predictor-variables-2" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="r-interlude-multiple-regression-with-2-predictor-variables-2">R INTERLUDE | multiple regression with 2 predictor variables</h2>
<ul>
<li>Now get rid of the interaction term, and set up a purely additive model</li>
<li>Did any of our estimates change? Why?</li>
<li>Did the degrees of freedom change? Why?</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>Add_lm <span class="ot">&lt;-</span> <span class="fu">lm</span>(y <span class="sc">~</span> g1<span class="sc">+</span>g2)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(Add_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Also try different combinations of genes</li>
<li>Finally, see if you can make a partial regression plot of just one predictor and the response variable</li>
</ul>
</section>
<section id="r-interlude-adding-a-polynomial" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="r-interlude-adding-a-polynomial">R INTERLUDE | adding a polynomial</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>Poly_lm_1 <span class="ot">&lt;-</span> <span class="fu">lm</span>(y <span class="sc">~</span> <span class="fu">poly</span>(g1, <span class="dv">1</span>))</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(Poly_lm_1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = y ~ poly(g1, 1))

Residuals:
     Min       1Q   Median       3Q      Max 
-0.50579 -0.18804 -0.03178  0.20893  0.54981 

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)  7.47450    0.06402  116.75   &lt;2e-16 ***
poly(g1, 1)  9.58694    0.28631   33.48   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.2863 on 18 degrees of freedom
Multiple R-squared:  0.9842,    Adjusted R-squared:  0.9833 
F-statistic:  1121 on 1 and 18 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
</section>
<section id="r-interlude-adding-a-polynomial-1" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="r-interlude-adding-a-polynomial-1">R INTERLUDE | adding a polynomial</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>Poly_lm_2 <span class="ot">&lt;-</span> <span class="fu">lm</span>(y <span class="sc">~</span> <span class="fu">poly</span>(g1, <span class="dv">2</span>))</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(Poly_lm_2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = y ~ poly(g1, 2))

Residuals:
     Min       1Q   Median       3Q      Max 
-0.48903 -0.19681 -0.01246  0.18486  0.54854 

Coefficients:
             Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)   7.47450    0.06575 113.681   &lt;2e-16 ***
poly(g1, 2)1  9.58694    0.29404  32.604   &lt;2e-16 ***
poly(g1, 2)2  0.07566    0.29404   0.257      0.8    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.294 on 17 degrees of freedom
Multiple R-squared:  0.9843,    Adjusted R-squared:  0.9824 
F-statistic: 531.5 on 2 and 17 DF,  p-value: 4.725e-16</code></pre>
</div>
</div>
</section>
<section id="r-interlude-adding-a-polynomial-2" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="r-interlude-adding-a-polynomial-2">R INTERLUDE | adding a polynomial</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>Poly_lm_3 <span class="ot">&lt;-</span> <span class="fu">lm</span>(y <span class="sc">~</span> <span class="fu">poly</span>(g1, <span class="dv">3</span>))</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(Poly_lm_3)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = y ~ poly(g1, 3))

Residuals:
    Min      1Q  Median      3Q     Max 
-0.5201 -0.1758 -0.0155  0.1628  0.5170 

Coefficients:
             Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)   7.47450    0.06720 111.233  &lt; 2e-16 ***
poly(g1, 3)1  9.58694    0.30051  31.902 6.51e-16 ***
poly(g1, 3)2  0.07566    0.30051   0.252    0.804    
poly(g1, 3)3  0.15773    0.30051   0.525    0.607    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.3005 on 16 degrees of freedom
Multiple R-squared:  0.9845,    Adjusted R-squared:  0.9816 
F-statistic: 339.4 on 3 and 16 DF,  p-value: 1.089e-14</code></pre>
</div>
</div>
</section>
<section id="r-interlude-adding-a-polynomial-3" class="level2 smaller">
<h2 class="smaller anchored" data-anchor-id="r-interlude-adding-a-polynomial-3">R INTERLUDE | adding a polynomial</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>Poly_lm_10 <span class="ot">&lt;-</span> <span class="fu">lm</span>(y <span class="sc">~</span> <span class="fu">poly</span>(g1, <span class="dv">10</span>))</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(Poly_lm_10)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = y ~ poly(g1, 10))

Residuals:
     Min       1Q   Median       3Q      Max 
-0.33720 -0.10156  0.00083  0.11606  0.30332 

Coefficients:
                Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)     7.474500   0.053607 139.431 2.55e-16 ***
poly(g1, 10)1   9.586936   0.239738  39.989 1.90e-11 ***
poly(g1, 10)2   0.075656   0.239738   0.316   0.7595    
poly(g1, 10)3   0.157731   0.239738   0.658   0.5271    
poly(g1, 10)4  -0.065839   0.239738  -0.275   0.7898    
poly(g1, 10)5   0.652130   0.239738   2.720   0.0236 *  
poly(g1, 10)6   0.167389   0.239738   0.698   0.5027    
poly(g1, 10)7  -0.349566   0.239738  -1.458   0.1788    
poly(g1, 10)8  -0.007755   0.239738  -0.032   0.9749    
poly(g1, 10)9   0.036730   0.239738   0.153   0.8816    
poly(g1, 10)10 -0.588592   0.239738  -2.455   0.0364 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.2397 on 9 degrees of freedom
Multiple R-squared:  0.9945,    Adjusted R-squared:  0.9883 
F-statistic: 161.6 on 10 and 9 DF,  p-value: 6.533e-09</code></pre>
</div>
</div>
</section>
</section>
<section id="model-selection-when-you-have-many-predictor-variables" class="level1">
<h1>Model selection when you have many predictor variables</h1>
<section id="model-selection" class="level2">
<h2 class="anchored" data-anchor-id="model-selection">Model selection</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/images_5b.004.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="model-selection-the-problems" class="level2">
<h2 class="anchored" data-anchor-id="model-selection-the-problems">Model selection | the problems</h2>
<ul>
<li><p>How to decide the complexity of polynomial: straight line regression, quadratic, cubic, ….</p></li>
<li><p>Which variables to keep/ discard when building a multiple regression model?</p></li>
<li><p>Selecting from candidate models representing different biological processes.</p></li>
</ul>
</section>
<section id="model-selection-a-beetle-example" class="level2">
<h2 class="anchored" data-anchor-id="model-selection-a-beetle-example">Model selection | a beetle example</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/images_5b.005.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:80.0%"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="start-with-linear-regression" class="level2">
<h2 class="anchored" data-anchor-id="start-with-linear-regression">Start with linear regression</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/images_5b.006.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:80.0%"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="quadratic-2nd-degree-polynomial" class="level2">
<h2 class="anchored" data-anchor-id="quadratic-2nd-degree-polynomial">Quadratic (2nd degree) polynomial?</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/images_5b.007.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:80.0%"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="quadratic-3rd-degree-polynomial" class="level2">
<h2 class="anchored" data-anchor-id="quadratic-3rd-degree-polynomial">Quadratic (3rd degree) polynomial?</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/images_5b.008.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:80.0%"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="a-polynomial-of-degree-5" class="level2">
<h2 class="anchored" data-anchor-id="a-polynomial-of-degree-5">A polynomial of degree 5?</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/images_5b.009.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:80.0%"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="a-polynomial-of-degree-10" class="level2">
<h2 class="anchored" data-anchor-id="a-polynomial-of-degree-10">A polynomial of degree 10?</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/images_5b.010.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:80.0%"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="the-problem-with-this-approach" class="level2">
<h2 class="anchored" data-anchor-id="the-problem-with-this-approach">The problem with this approach</h2>
<ul>
<li>The log likelihood of the model increases with the number of parameters</li>
<li>So does <span class="math inline">\(r^2\)</span></li>
<li>Isn’t this good - the best fit to the data?</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/images_5b.011.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:70.0%"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="the-problem-with-this-approach-1" class="level2">
<h2 class="anchored" data-anchor-id="the-problem-with-this-approach-1">The problem with this approach</h2>
<ul>
<li>Does it violate a principle of parsimony</li>
<li>Fit no more parameters than is necessary.</li>
<li>If two or more models fit the data almost equally well, prefer the simpler model.</li>
</ul>
<blockquote class="blockquote">
<p>“models should be pared down until they are minimal and adequate”</p>
</blockquote>
<p>Crawley 2007, p325</p>
</section>
<section id="lets-consider-our-objectives" class="level2">
<h2 class="anchored" data-anchor-id="lets-consider-our-objectives">Let’s consider our objectives</h2>
<ul>
<li><p>Model should predicts well</p></li>
<li><p>Approximates true relationship between the variables</p></li>
<li><p>Be able to evaluate a wider array of models. Not only or more “reduced” models.</p></li>
<li><p>But an overfitted model may not predict well in the future!</p></li>
</ul>
</section>
<section id="a-polynomial-of-degree-10-1" class="level2">
<h2 class="anchored" data-anchor-id="a-polynomial-of-degree-10-1">A polynomial of degree 10?</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/images_5b.010.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:80.0%"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="section-3" class="level2 flexbox vcenter">
<h2 class="flexbox vcenter anchored" data-anchor-id="section-3"></h2>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>