---
title: "BioE_Stats_2025_HW1 - EDA, t-test and Power Calculations"
format: html
editor: visual
---

## Directions

**Assignment:** Your task is to use R and Quarto Markdown to write a short report including the following tasks. You will use both the **microbiota/RNAseq data** or the **knee injury** dataset that you have been using in class. The goal is for you to create a structured Quarto Markdown document that has a series of code blocks, and is well annotated with respect to the steps that you are performing and your findings. I provide the rough steps below, but you can depend on the class material and the web to augment and expand the document.

**Due:** Submit your work via Canvas by the end of the day (midnight) on **Wednesday, January 21st.** Please submit **both the Quarto_md file *and* the resulting html or pdf file**. You can work with other members of class, but I expect each of you to construct and run all of the scripts yourself.

## Problem 1 - Microbiota/gene expression Exploratory Data Analysis

### Edit the YAML header and create a set up code block

-   Edit the YAML header to have the proper title and author, as well as any other global options you'd like.
-   Create a `setup` code block that has any necessary libaries loaded.

### Read in the data

-   Read in the files
-   Use `str`, `head` and `glimpse` to get an idea of the data structure
-   Create a `tibble` data object using this dataset (the function is `as_tibble`)
-   Create a table that functions as a `data dictionary` with your best guess as to what each of the variables is (of course for the gene expression variables you just need to describe them once)

### Perform some wrangling

-   For the microbiota data, `Select` all of the categorical variables and only 3 of the continuous gene expression variables and assign them to a new data object.
-   `Mutate` some of the numeric variables by performing a square root transformation making a new variable for each of the original (keep all six variables in the dataset).
-   `Summarise` the mean and standard deviation for each of the gene count variables grouped by the ‘Microbiota’ and ‘Genotype’ categorical variables (hint, using `group_by` may make this easier; also remember your `na.rm=TRUE` to remove the `na` values)

### Graphical Communication with `ggplot2`

Insert a code chunk and complete the following tasks using your gene expression tibble:

-   Create a barplot of microbiota and genotype- what does this tell you about experimental design?

-   Create a histogram and frequency polygram of the expression of one particular gene.

-   Create a boxplot of the expression of one gene by microbiota/genotype combination. Reorder it by median expression and flip the coordinates.

-   Explore the coexpression patterns among your genes by plotting a couple pairwise combinations using `geom_point`. Plot your most interesting coexpression by treatment using `facet_wrap`.

## Problem 2 - Knee injury Exploratory Data Analysis

- For the second problem I would like you to follow the general steps above and perform an EDA on the `knee injury` data set.

- You won't be following the exact same steps, but you still would like to know about the general structure of the data.

- The goal of this problem is to apply and extend your knowledge to a new data set.

