---
title: "Week 1 Statistics for Bioengineering"
author: "Bill Cresko"
format: 
  revealjs:
    footer: BioE_Stats_2026 - Knight Campus 
    transition: fade
    transition-speed: slow
editor: visual
---

```{r}
library(tidyverse)
library(gt)
library(readxl)
theme_set(theme_minimal())
```

## Statistics for Bioengineers and Bioscientists

![](images/Cover2.tiff){width="100"}

## Goals of the course

::: {.callout-tip title="This is a practical course and we will learn by doing"}
-   Teach fundamental skills for your scientific careers
-   Provide a broad coverage of the core components of modern statistics
-   Provide you with the computational tools necessary to carry out your work
-   To prepare you for more advanced statistics and programming education
:::

## What will you learn? {.smaller}

::: incremental
-   Read and write code in Unix and R
-   Implement reproducible research practices
-   Exploratory data analysis and visualization
-   Probability in the context of distributions and sampling
-   Experimental design
-   p-values, test statistics, and types of errors
-   Statistical analyses such as t-tests and contigency tests
-   Linear and non-linear modeling
-   Classical machine learning
-   Ordination such as PCA, tSNE and UMAP
:::

------------------------------------------------------------------------

## 

```{r, echo=FALSE, out.width='80%', fig.align='center'}
knitr::include_graphics("images/week_01.001.jpeg")
```

## 

```{r, echo=FALSE, out.width='100%', fig.align='center'}
knitr::include_graphics("images/week_01.002.jpeg")
```

## 

```{r, echo=FALSE, out.width='100%', fig.align='center'}
knitr::include_graphics("images/week_01.003.jpeg")
```

## 

```{r, echo=FALSE, out.width='100%'}
knitr::include_graphics("images/week_01.004.jpeg")
```

------------------------------------------------------------------------

## Class Logistics {.smaller}

-   Meet Tuesdays and Thursdays from 4:00pm - 5:50pm in KC107
    -   Most of class time will be hands-on coding practice, less time lecturing
-   Coding and statistics practice via homework in Weeks 2-9
    -   Available Tuesday of that week, due before class on Tuesday in two weeks
-   Weeks 10-11 you will complete a final coding project
    -   Design script(s) that works with your research and interests using the skills you've learned this term

## Required Materials {.smaller}

-   No textbooks or purchases required
-   Access to a laptop or computer running Windows, MacOS, or Linux operating systems
-   Announcements and assignments posted on Canvas
-   The majority of course material on our class website https://wcresko.github.io/BioE_Stats_Wi2026/
-   The reading material can be found here: https://wcresko.github.io/BioE_Stats_Book/

## Class Introductions

::: {.callout-tip title="Who are you?"}
-   Your name
-   Year in grad school
-   Home lab or rotation lab
-   What is your good news this week?
-   What has your experience with programming/statistics been like?
:::


# Statistics&Computers = Chocolate&Peanut Butter

## Why do we need statistics? {.smaller}

-   We almost never know the world perfectly, but still want to draw conclusions or make decisions
-   We need to estimate underlying parameters from samples of data
-   Sometimes we need to test hypotheses using data
-   Other times we need to more succinctly summarize and/or visualize large amounts of data
-   There are well known mathematical rules that help us
-   Statistics can be done by hand, but computers let us do most of the mathematics quickly

## Why do we need statistics? {.smaller}

-   We want to turn data into conclusions about the world
    -   point estimates and confidence intervals
    -   experimental design
    -   hypothesis testing
    -   data reduction of highly dimensional data
-   We need a firm understanding of probability, sampling and distributions

## Why do we need coding and scripting? {.smaller}

-   It is incredibly fast and powerful, particularly for repeated actions
-   It allows you to do thousands of ‘clicks’ with single commands
-   Ability to analyze large datasets that Excel and other GUIs can't handle well
-   Access to thousands of free programs made for and by scientists
-   The commands work almost identically across platforms
-   Ability to use computer clusters like Talapas
-   ***REPRODUCIBILITY***

## What is the difference between coding and scripting? {.smaller}

-   `coding` generally involves computer languages that use `compilers`
    -   C\^{++}, Fortran, etc
-   `scripting` generally involves computer languages that are `interpreted` on the fly
    -   Python, R, Julia, etc.
-   coding - faster but less flexible; scripting - flexible but slower
-   The distinction between the two has become somewhat fuzzy and most modern analytical pipelines contain a combination of both

# Installing Programs

## 

::: callout-note
-   Mac and Linux systems run using the same language, but previous versions of Windows lacks some of the basic features found on other systems
-   To help you practice and learn how to code in Unix, we will help you install some programs on your computer for running Unix
-   R and RStudio should work on any computer
:::

## Running `R` {.smaller}

-   Need to make sure that you have `R` installed
    -   locally or on a server
    -   https://www.r-project.org
-   Can run R from the command line
    -   just type `R`
    -   can run it locally as well as on clusters

## Running `R`

-   Install an R *Integrated Development Environment* (IDE)
    -   RStudio: http://www.rstudio.com
    -   Makes working with R much easier, particularly for a new R user
    -   Run on Windows, Mac or Linux OS
    -   Use the RStudio Desktop Open Source (Free)

## `RStudio` {.smaller}

```{r, echo=FALSE, out.width='75%', fig.align='center'}
knitr::include_graphics("images/RStudio_Screenshot.png")
```

## Accessing the shell {.smaller}

-   Mac users: open the "Terminal" app, or use another app like 'iTerm2'

-   Linux users: open one of several "Terminal" apps

    ::: callout-caution
    ## Windows users have a little more work to do

    See the next slides
    :::

## Accessing the shell {.smaller}

### *Windows users*

-   Guide: https://ubuntu.com/tutorials/install-ubuntu-on-wsl2-on-windows-10#1-overview
-   Run Windows PowerShell as administrator
-   Install WSL2 by typing `wsl --install`
-   Restart your computer
-   Search for and install Ubuntu from Microsoft store app
-   ***OR*** type `wsl --install -d ubuntu` on PowerShell to do both at once

## *Accessing the shell on Windows*

-   Open Ubuntu and set up a username and password
-   Does not have to match your login info for Windows
-   Run `sudo apt update` then `sudo apt upgrade` to ensure everything is up to date
-   Will need to create folders and files within your Ubuntu folder on your computer

# BREAK

# Computational Tools for Coding and Scripting

## What is a Computer? {.smaller}

-   At its core, a computer operates using combinations of 0s and 1s (binary)
-   Base programming languages translate binary into short commands (e.g., Fortran, C)
-   Programming languages have been built on top of one another until we get English-like commands (Unix, R, Python, etc.)
-   System of directories (folders) and files
-   `/` = the root directory, which holds all other directories

::: callout-note
Navigation in programming consists of jumping between directories and telling your computer where files are located
:::

## A Note on Programming Languages {.smaller}

-   Programming languages are called *languages* for a reason
-   Learning to program is a lot like learning a new spoken language
-   Each programming language is different, but learning one will help you learn another
-   **Use it or lose it!** Practice is the best way to learn and maintain your coding skills

::: callout-tip
Computers are literal! You have to use very explicit and pre-defined phrases to tell your computer exactly what you want
:::


# The Command Line {.flexbox .vcenter}

```{r, echo=FALSE, out.width='90%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/w1_code.jpeg")
```

## What is Unix? {.smaller}

::: incremental
-   A scripting language developed in 1969, released in 1973
-   Serves as the base language for many programs and computers
-   Is the operating system for computers
-   Linux is an open-source version of the same language
:::

## What is a shell? {.flexbox .vcenter data-background="images/w1_shell.jpeg"}

-   The ‘shell’ is a program that runs UNIX and takes in commands and gives them to the operating system

-   Bash acts as the shell in macs, linux, and now windows

-   You can access the shell via a terminal window

##  {.flexbox .vcenter data-transition="fade-in"}

```{r, echo=FALSE, out.width='90%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/week_01.010.jpeg")
```

## Recipes for a shell command {.smaller}

-   Prompt: notation used to indicate your computer is ready to accept a new command
-   Command: the building blocks of programming, tell computer to do a specific task
-   Options: change the behavior of a command
-   Argument: what the command should operate on

```{r, echo=FALSE, out.width='50%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/w1_shell_command.jpeg")
```

## Why use the shell? {.smaller}

-   Speed
-   Examine large and/or unique files
-   Access super computers
-   Use programs only available via shell
-   The commands work almost identically across platforms
-   You can even use them on a large computer cluster like Talapas
-   It is incredibly powerful particularly for repeated actions
-   It allows you to do thousands of ‘clicks’ with single commands
-   ***THIS IS HOW WE KNOW TO RUN OUR SCRIPTS***

## Where do you get help? {.smaller}

-   Manual pages!
    -   The shell has manuals for all basic commands
    -   Type `man [command_name]` to access the manual for a specific command
    -   Type `q` to exit
-   Also...the internet!

# Common navigation commands

## The way you normally navigate

```{r, echo=FALSE, out.width='90%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/w1_finder.jpeg")
```

##  {.flexbox .vcenter data-transition="fade-in"}

```{r, echo=FALSE, out.width='90%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/w1_navigate.jpeg")
```

## How is a computer organized? {.smaller}

-   System of directories (folders) and files
-   / = the root directory, which holds all other directories
-   Most of your files will be located under /Users in a directory of your username
-   the `~` is shorthand for your home folder
-   Navigation in the shell consists of jumping up and down between directories and seeing what’s in them
-   The "path" refers to the location a file is in
    -   ex: "/Users/wcresko/Documents"

```{r, echo=FALSE, out.width='100%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/w1_file_structure.jpeg")
```

##  {.flexbox .vcenter}

```{r, echo=FALSE, out.width='90%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/week_01.012.jpeg")
```

##  {.flexbox .vcenter}

```{r, echo=FALSE, out.width='90%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/week_01.013.jpeg")
```

##  {.flexbox .vcenter}

```{r, echo=FALSE, out.width='90%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/week_01.015.jpeg")
```

##  {.flexbox .vcenter}

```{r, echo=FALSE, out.width='90%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/week_01.016.jpeg")
```

##  {.flexbox .vcenter}

```{r, echo=FALSE, out.width='90%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/week_01.017.jpeg")
```

##  {.flexbox .vcenter}

```{r, echo=FALSE, out.width='90%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/week_01.018.jpeg")
```

## Common navigation commands {.smaller}

-   **`pwd`** = “print working directory”, which will print where you currently are in the system
    -   In Windows, **`cd`** will print your working directory
-   **`ls`** = “list”, list all directories and files in your current position
    -   `ls –F` = denote which results are directories, files, etc.
    -   `ls –l` = ”long format”, lists total file sizes
    -   `ls –r` = “reverse”, lists the results in reverse order
    -   `ls –S` = “size”, sort results by size
    -   `ls –t` = “time”, sort results by time created, from most recent to last
-   **`cd`** = “change directory”, will place you in a new position based on your path argument
    -   `cd ..` = go up one directory
    -   `cd –` = go to the directory you were at last (like the back arrow on an internet browser)

## Let's practice! {.smaller}

-   Try navigating around your computer using `cd` and `ls`
-   If you are on Ubuntu, you may need to create some empty directories in your Ubuntu folder before navigating in the terminal

# Working with files

## Making new files {.smaller}

-   Make new folders: `mkdir`
-   Make new files: `nano`, `touch`
-   Rename files: `mv`
-   Move files: `mv`
-   Copy files: `cp`
-   Delete files: `rm`
-   Examining file length: `wc`
-   Reading files: `cat`
-   Looking at beginning or end: `head` or `tail`

## Things to keep in mind {.smaller}

-   The shell trusts you
    -   It will delete files you say to delete
    -   It will override files if you name 2 things the same
-   Naming conventions
    -   Avoid spaces
    -   Don’t start with a –
    -   Stick to letters, numbers, . , -, and \_
-   Use appropriate file extensions in file names
    -   Some software expect files with certain extensions (.fasta, .txt, etc.)

## Let's try it out {.smaller}

-   Make a new directory called whatever you'd like.
    -   Add a file named `practice.txt` to the directory and add some text to it
    -   Read the contents of the file and get its length
    -   Rename the file to `super_practice.txt`
-   Move the file to a new folder named `sandbox`
    -   Make a copy of the file named `super_practice_copy.txt`
-   Read the contents of the file and get its length to make sure it’s the same as Super_practice.txt
-   Delete the original `super_practice.txt`

## Break

# Tidy Data

## An example to get us started {.smaller}

##  {.flexbox .vcenter}

```{r, echo=FALSE, out.width='100%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/week_01.005.jpeg")
```

##  {.flexbox .vcenter}

```{r, echo=FALSE, out.width='90%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/week_01.006.jpeg")
```

## Data set rules of thumb (aka Tidy Data) {.smaller}

-   Store a copy of data in nonproprietary software and hardware formats, such as plain ASCII text (aka a flat file)
-   Leave an uncorrected file when doing analyses
-   Use descriptive names for your data files and variables
-   Include a header line with descriptive variable names
-   Maintain effective metadata about the data
-   When you add observations to a database, add rows
-   When you add variables to a database, add columns, not rows
-   A column of data should contain only one data type

## Common Data Types {.smaller}

| Statistical Term | Description | R Data Type |
|:----------------------------|:---------------------|:---------------------|
| Numerical (continuous) | Measurements that can take any value (e.g., height, weight) | `numeric` / `double` |
| Numerical (discrete) | Whole number counts (e.g., number of cells) | `integer` |
| Categorical (nominal) | Unordered categories (e.g., species, color) | `factor` or `character` |
| Categorical (ordinal) | Ordered categories (e.g., low/medium/high) | `ordered factor` |
| Logical / Boolean | TRUE or FALSE values | `logical` |
| Text / String | Free-form text data | `character` |
| Date / Time | Temporal data | `Date`, `POSIXct` |

::: callout-tip
Use `class()` or `str()` in R to check the data type of any variable
:::

##  {.flexbox .vcenter}

```{r, echo=FALSE, out.width='90%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/week_01.007.jpeg")
```

------------------------------------------------------------------------

## Computational Tools - R and RStudio {.flexbox .vcenter}

```{r, echo=FALSE, out.width='75%', fig.align='center'}
knitr::include_graphics("images/RStudio_Screenshot.png")
```

## The RStudio Interface {.smaller}

RStudio has four main panes that help organize your workflow:

| Pane | Location | Purpose |
|:------------------|:---------------------------|:-------------------------|
| **Source** | Top-left | Write and edit scripts, markdown files, and view data |
| **Console** | Bottom-left | Interactive R session, run commands directly |
| **Environment/History** | Top-right | View objects in memory, command history |
| **Files/Plots/Packages/Help** | Bottom-right | Navigate files, view plots, manage packages, get help |

::: callout-tip
You can customize the pane layout in Tools → Global Options → Pane Layout
:::

## Why use `R`? {.flexbox .vcenter .smaller}

::: incremental
-   Good general scripting tool for statistics and mathematics
-   Powerful and flexible and free
-   Runs on all computer platforms
-   New enhancements coming out all the time
-   Superb data management & graphics capabilities
:::

## Why use `R`? {.flexbox .vcenter .smaller}

::: incremental
-   Reproducibility - can keep your scripts to see exactly what was done
-   You can write your own functions
-   Lots of online help available
-   Can use a nice GUI front end such as `Rstudio`
-   Can embed your `R` analyses in dynamic, polished files using `Markdown`
-   Markdown can be reused for websites, papers, books, presentations...
:::

## `R` Resources {.smaller}

Helpful links for learning and troubleshooting R:

| Resource | URL |
|:---|:---|
| The R Project | <http://www.r-project.org> |
| RStudio | <http://www.rstudio.com> |
| R for Data Science (free book) | <https://r4ds.had.co.nz> |
| Quick R Homepage | <http://www.statmethods.net> |
| Bioconductor | <http://www.bioconductor.org> |
| Stack Overflow | <https://stackoverflow.com/questions/tagged/r> |

::: callout-tip
When you get an error, copy and paste it into Google — chances are someone else has had the same problem!
:::

## Installing and Loading Packages {.smaller}

-   R's power comes from thousands of community-contributed packages
-   **Install once** (downloads package to your computer)
-   **Load each session** (makes functions available)

```{r, eval = FALSE, echo = TRUE}
# Install a package (only need to do this once)
install.packages("tidyverse")

# Load the package (do this each R session)
library(tidyverse)
```

::: callout-note
You can also install packages via RStudio: Tools → Install Packages
:::

## The Tidyverse Ecosystem {.smaller}

The `tidyverse` is a collection of R packages designed for data science that share a common philosophy and syntax:

| Package   | Purpose                                               |
|:----------|:------------------------------------------------------|
| `dplyr`   | Data manipulation (filter, select, mutate, summarize) |
| `ggplot2` | Data visualization                                    |
| `tidyr`   | Reshaping data (pivot, separate, unite)               |
| `readr`   | Fast reading of rectangular data (CSV, TSV)           |
| `stringr` | String manipulation                                   |
| `purrr`   | Functional programming tools                          |

Loading `tidyverse` loads all of these at once!


# Writing Scripts in R

## Rscript basics

-   A series of R commands that will be executed
-   Can add comments using hashtags `#`
-   Can have pipes to connect one step to the next


## BASICS of `R` {.flexbox .vcenter .smaller}

-   Commands can be submitted through the terminal, console or scripts
-   In your scripts, anything that follows '\#' symbol (aka hash) is just for humans
-   Notice on these slides I'm evaluating the code chunks and showing output
-   The output is shown here after the two `#` symbols and the number of output items is in `[]`
-   Also notice that `R` follows the normal priority of mathematical evaluation

```{r basic multiplication, echo=TRUE}
4*4
```

```{r more multiplication, echo=TRUE}
(4+3*2^2)
```

## Assigning Variables {.flexbox .vcenter}

-   A better way to do this is to assign variables
-   Variables are assigned values using the `<-` operator.
-   Variable names must begin with a letter, but other than that, just about anything goes.
-   Do keep in mind that `R` is case sensitive.

## Assigning Variables {.flexbox .vcenter}

```{r assigning variables, echo=TRUE, tidy=TRUE}

x <- 2
x*3
y <- x * 3
y-2

```

These do not work

```{r eval=FALSE, echo=TRUE}
3y <- 3
3*y <- 3
```

## Arithmetic operations on functions {.smaller}

-   Arithmetic operations can be performed easily on functions as well as numbers.
-   Try the following, and then your own.

```{r functions, eval = FALSE, echo = TRUE}
x+2
x^2
log(x)
```

-   Note that the last of these - `log` - is a built in function of `R`, and therefore the object of the function needs to be put in parentheses
-   These parentheses will be important, and we'll come back to them later when we add arguments after the object in the parentheses\
-   The outcome of calculations can be assigned to new variables as well, and the results can be checked using the 'print' command

## Arithmetic operations on functions {.flexbox .vcenter}

```{r, echo=TRUE}
y <- 67
print(y)

x <- 124
z <- (x*y)^2
print(z)
```

## STRINGS {.flexbox .vcenter}

-   Variables and operations can be performed on characters as well
-   Note that characters need to be set off by quotation marks to differentiate them from numbers
-   The `c` stands for `concatenate`
-   Note that we are using the same variable names as we did previously, which means that we're overwriting our previous assignment
-   A good rule of thumb is to use new names for each variable, and make them short but still descriptive

## STRINGS {.flexbox .vcenter}

```{r, echo=TRUE}
x <- "I Love"
print (x)
y <- "Biostatistics"
print (y)
z <- c(x,y)
print (z)
```

## FACTORS {.flexbox .vcenter}

-   The variable `z` is now what is called a list of character values.
-   Sometimes we would like to treat the characters as if they were units for subsequent calculations.
-   These are called `factors`, and we can redefine our character variables as factors.
-   This might seem a bit strange, but it’s important for statistical analyses where we might want to see the mean or variance for two different treatments.

## FACTORS {.flexbox .vcenter}

```{r, eval = FALSE, echo = TRUE}
z_factor <- as.factor(z)
print (z_factor)
```

-   Note that factor levels are reported alphabetically

## Data Type Coercion {.smaller}

You can convert between data types using `as.*` functions:

```{r, eval = FALSE, echo = TRUE}
# Convert to numeric
as.numeric("42")        # Returns: 42
as.numeric("hello")     # Returns: NA (with warning)

# Convert to character
as.character(42)        # Returns: "42"
as.character(TRUE)      # Returns: "TRUE"

# Convert to factor
as.factor(c("low", "medium", "high"))

# Convert to logical
as.logical(1)           # Returns: TRUE
as.logical(0)           # Returns: FALSE
```

::: callout-warning
Be careful! `as.numeric()` on a factor returns the underlying integer codes, not the labels. Use `as.numeric(as.character(x))` instead.
:::

## VECTORS {.flexbox .vcenter}

-   In general `R` thinks in terms of vectors (a list of characters, factors or numerical values) and it will benefit any `R` user to try to write programs with that in mind, as it will simplify most things.
-   Vectors can be assigned directly using the 'c()' function and then entering the exact values.

## VECTORS {.flexbox .vcenter}

```{r Example vectors, echo=TRUE}
x <- c(2,3,4,2,1,2,4,5,10,8,9)
print(x)
```

## Basic Statistics {.flexbox .vcenter}

-   Many functions exist to operate on vectors.
-   Combine these with your previous variable to see what happens.
-   Also, try to find other functions (e.g. standard deviation).

## Basic Statistics {.flexbox .vcenter}

```{r, eval = FALSE, echo = TRUE}
mean(x)
median(x)
var(x)
log(x)      # natural log (base e)
log10(x)    # base 10 log
sqrt(x)
sum(x)
length(x)
sample(x, replace = T)
```

-   Notice that the last function (`sample`) has an argument (`replace=T`)
-   Arguments simply modify or direct the function in some way
-   There are many arguments for each function, some of which are defaults

## Logical Operators and Comparisons {.smaller}

Logical operations are essential for filtering and subsetting data:

| Operator   | Meaning               | Example             |
|:-----------|:----------------------|:--------------------|
| `==`       | equals                | `x == 5`            |
| `!=`       | not equals            | `x != 5`            |
| `<`, `>`   | less/greater than     | `x < 10`            |
| `<=`, `>=` | less/greater or equal | `x >= 0`            |
| `&`        | and                   | `x > 0 & x < 10`    |
| `|`        | or                    | `x < 0 | x > 10`    |
| `!`        | not                   | `!is.na(x)`         |
| `%in%`     | is element of         | `x %in% c(1, 2, 3)` |

```{r, eval = FALSE, echo = TRUE}
x <- c(1, 5, 10, 15, 20)
x > 10              # Returns: FALSE FALSE FALSE TRUE TRUE
x[x > 10]           # Returns: 15 20 (subsetting!)
```

## Handling Missing Data (NA) {.smaller}

Real data almost always has missing values. In R, these are represented as `NA`:

```{r, eval = FALSE, echo = TRUE}
x <- c(1, 2, NA, 4, 5)

# Checking for NA
is.na(x)            # Returns: FALSE FALSE TRUE FALSE FALSE
!is.na(x)           # Invert to find non-missing

# WARNING: This doesn't work!
x == NA             # Returns: NA NA NA NA NA (not what you want!)

# Handling NA in functions
mean(x)             # Returns: NA (can't compute)
mean(x, na.rm = TRUE)   # Returns: 3 (ignores NA)

# Removing NA values
na.omit(x)          # Returns vector without NA
```

::: callout-important
Always use `is.na()` to check for missing values, never `== NA`
:::

## Special Numeric Values {.smaller}

R has several special values for edge cases:

| Value  | Meaning                       | Example       |
|:-------|:------------------------------|:--------------|
| `NA`   | Not Available (missing data)  | `c(1, NA, 3)` |
| `NaN`  | Not a Number (undefined math) | `0/0`         |
| `Inf`  | Positive infinity             | `1/0`         |
| `-Inf` | Negative infinity             | `-1/0`        |

```{r, eval = FALSE, echo = TRUE}
# Testing for special values
is.na(NA)       # TRUE
is.nan(NaN)     # TRUE
is.infinite(Inf) # TRUE
is.finite(42)   # TRUE

# NaN is also considered NA
is.na(NaN)      # TRUE (!)
```

## Getting Help {.flexbox .vcenter}

-   Getting Help on any function is very easy - just type a question mark and the name of the function.
-   There are functions for just about anything within `R` and it is easy enough to write your own functions if none already exist to do what you want to do.
-   In general, function calls have a simple structure: a function name, a set of parentheses and an optional set of parameters to send to the function.
-   Help pages exist for all functions that, at a minimum, explain what parameters exist for the function.\
-   Help can be accessed a few ways - try them :

## Getting Help {.flexbox .vcenter}

```{r, eval = FALSE, echo = TRUE}
help(mean)
?mean
example(mean)
help.search("mean")
apropos("mean")
args(mean)
```

## Common R Errors (Don't Panic!) {.smaller}

You will see errors — everyone does! Here's how to read them:

| Error Message | What It Means | How to Fix |
|:-------------------------|:-------------------------|:--------------------|
| `object 'x' not found` | Variable doesn't exist | Check spelling, or run the line that creates it |
| `could not find function` | Package not loaded or typo | Load package with `library()` or check spelling |
| `unexpected ')'` or `'}'` | Mismatched parentheses/brackets | Count your opening and closing brackets |
| `non-numeric argument` | Wrong data type | Check with `class()`, convert if needed |
| `subscript out of bounds` | Index doesn't exist | Vector/list is shorter than you think |

::: callout-tip
Read error messages from the bottom up — the last line usually tells you what went wrong
:::

## Creating vectors {.smaller}

-   Creating vector of new data by entering it by hand can be a drag
-   However, it is also very easy to use functions such as `seq` and `sample`
-   Try the examples below Can you figure out what the three arguments in the parentheses mean?
-   Try varying the arguments to see what happens.
-   Don't go too crazy with the last one or your computer might slow way down

## Creating vectors {.smaller}

```{r, echo = TRUE}
seq_1 <- seq(0.0, 10.0, by = 0.1)
print(seq_1)
seq_2 <- seq(10.0, 0.0, by = -0.1)
print(seq_2)
```

## Creating vectors {.smaller}

```{r, echo = TRUE}
seq_square <- (seq_2)*(seq_2)
print(seq_square)
```

## Creating vectors {.smaller}

```{r, echo = TRUE}
seq_square_new <- (seq_2)^2
print(seq_square_new)
```

## Drawing samples from distributions {.flexbox .vcenter}

-   Here is a way to create your own data sets that are random samples.
-   Again, play around with the arguments in the parentheses to see what happens.

## Drawing samples from distributions {.smaller}

```{r Samples from distributions 1, out.width='100%', echo = TRUE}
x <- rnorm (10000, 0, 10)
y <- sample (1:10000, 10000, replace = T)
xy <- cbind(x,y)
plot(x,y) 
```

## Drawing samples from distributions {.smaller}

```{r Samples from distributions 2, out.width='100%', echo = TRUE}
x <- rnorm (10000, 0, 10)
y <- sample (1:10000, 10000, replace = T)
xy <- cbind(x,y)
plot(xy)
```

## Drawing samples from distributions {.smaller}

```{r Samples from distributions 3, out.width='100%', echo = TRUE}
x <- rnorm (10000, 0, 10)
y <- sample (1:10000, 10000, replace = T)
xy <- cbind(x,y)
hist(x)
```

## Drawing samples from distributions {.flexbox .vcenter}

-   You’ve probably figured out that y from the last example is drawing numbers with equal probability.
-   What if you want to draw from a distribution?
-   Again, play around with the arguments in the parentheses to see what happens.

## Drawing samples from distributions {.flexbox .vcenter out.width="50%"}

```{r, out.width='100%', echo = TRUE}
x <-rnorm(1000, 0, 100)
hist(x, xlim = c(-500,500))
curve(50000*dnorm(x, 0, 100), xlim = c(-500,500), add=TRUE, col='Red')
```

-   `dnorm()` generates the probability density, which can be plotted using the `curve()` function.
-   Note that is curve is added to the plot using `add=TRUE`

## Visualizing Data {.flexbox .vcenter}

-   So far you've been visualizing just the list of output numbers
-   Except for the last example where I snuck in a `hist` function.
-   You can also visualize all of the variables that you've created using the `plot` function (as well as a number of more sophisticated plotting functions).
-   Each of these is called a `high level` plotting function, which sets the stage
-   `Low level` plotting functions will tweak the plots and make them beautiful

## Visualizing Data {.flexbox .vcenter}

-   What do you think that each of the arguments means for the plot function?
-   A cool thing about `R` is that the options for the arguments make sense.
-   Try adjusting an argument and see if it works
-   Note next week we will be exploring the plotting in `GGPlot2`

## Visualizing Data {.flexbox .vcenter .smaller}

```{r, echo = TRUE}
seq_1 <- seq(0.0, 10.0, by = 0.1) 
plot (seq_1, xlab="space", ylab ="function of space", type = "p", col = "red")
```

## Putting plots in a single figure {.flexbox .vcenter}

-   On the next slide
-   The first line of the lower script tells R that you are going to create a composite figure that has two rows and two columns. Can you tell how?
-   Now, modify the code to add two more variables and add one more row of two panels.

```{r, echo=TRUE, outwidth="50%", out.height="20%"}
seq_1 <- seq(0.0, 10.0, by = 0.1)
seq_2 <- seq(10.0, 0.0, by = -0.1)
```

## Putting plots in a single figure {.flexbox .vcenter .smaller}

```{r, echo=TRUE, outwidth="50%", out.height="50%"}
par(mfrow=c(2,2))
plot (seq_1, xlab="time", ylab ="p in population 1", type = "p", col = 'red')
plot (seq_2, xlab="time", ylab ="p in population 2", type = "p", col = 'green')
plot (seq_square, xlab="time", ylab ="p2 in population 2", type = "p", col = 'blue')
plot (seq_square_new, xlab="time", ylab ="p in population 1", type = "l", col = 'yellow')
```

## Example using binomial distribution {.flexbox .vcenter}

-   As above for the normal distribution, data can be generated by being sampled from nearly any distribution and then visualized.
-   Below I’m having you use the ‘histogram’ function. What does it do?

## Example using binomial distribution {.smaller}

-   10 successes (out of 20 trials) is the most frequent outcome

```{r binomial function, echo=TRUE}
heads <- rbinom(n=1000, size=20, prob=0.5)
hist(heads)
```

## Example using binomial distribution {.smaller}

-   This kind of statement can be run in one line as well, which is sometimes easier.

```{r binomial function again, echo=TRUE}
hist(rbinom(n=1000, size=20, prob=0.5))
```

## Creating Data Frames in R {.smaller}

-   As you have seen, in R you can generate your own random data set drawn from nearly any distribution very easily.
-   Often we will want to use collected data.
-   Now, let’s make a dummy dataset to get used to dealing with data frames
-   Set up three variables (hydrogel_concentration, compression and conductivity) as vectors

```{r, echo=TRUE}
hydrogel_concentration <- factor(c("low", "high", "high", "high", "medium", "medium", "medium","low"))
compression <- c(3.4, 3.4, 8.4, 3, 5.6, 8.1, 8.3, 4.5)
conductivity <- c(0, 9.2, 3.8, 5, 5.6, 4.1, 7.1, 5.3)
```

-   Create a data frame where vectors become columns

```{r, echo=TRUE}
mydata <- data.frame(hydrogel_concentration, compression, conductivity)
row.names(mydata) <- c("Sample_1", "Sample_2", "Sample_3", "Sample_4", 
                       "Sample_5", "Sample_6", "Sample_7", "Sample_8")
```

-   Now you have a hand-made data frame with row names
-   Take a look at it in the data section of RStudio

## Setting Your Working Directory {.smaller}

The **working directory** is the folder R looks in by default when reading or writing files:

```{r, eval = FALSE, echo = TRUE}
# Check your current working directory
getwd()

# Set a new working directory
setwd("/path/to/your/folder")

# List files in working directory
list.files()
```

**In RStudio (easier method):**

-   Session → Set Working Directory → Choose Directory
-   Or use the Files pane: navigate to folder, then click "More" → "Set As Working Directory"

::: callout-tip
Using RStudio Projects (File → New Project) automatically sets your working directory and keeps your work organized
:::

## Common Data File Formats {.smaller}

| Format | Extension | Separator | Notes |
|:----------------|:------------------|:------------------|:----------------|
| Comma-separated | `.csv` | `,` | Most common, works with Excel |
| Tab-separated | `.tsv` or `.txt` | `\t` (tab) | Good for data with commas in text |
| Plain text | `.txt` | varies | Check the file to see delimiter |
| Excel | `.xlsx` | n/a | Use `readxl` package |

::: callout-warning
Always check your imported data! Use `head()`, `str()`, or `View()` to verify it loaded correctly.
:::

## Reading in Data Frames in R {.smaller}

-   A strength of `R` is being able to import data from an external source
-   The `sep` argument specifies the delimiter (comma, tab, etc.)
-   The `header` argument indicates whether the first row contains column names
-   Use `row.names` to specify which column contains row identifiers

```{r, eval = FALSE, echo = TRUE}
# Comma-separated file
data_csv <- read.csv('yourfile.csv')

# Tab-separated file  
data_tsv <- read.table('yourfile.txt', header=TRUE, sep='\t')

# Excel file (requires readxl package)
library(readxl)
data_xlsx <- read_excel('yourfile.xlsx')
```

## Exporting Data Frames in R {.smaller}

Save your data frames to files for sharing or later use:

```{r, eval = FALSE, echo = TRUE}
# Export as CSV
write.csv(YourData, "output.csv", row.names = FALSE)

# Export as tab-separated
write.table(YourData, "output.txt", sep = "\t", row.names = FALSE, quote = FALSE)

# Using tidyverse (readr package)
write_csv(YourData, "output.csv")
```

## Indexing in Data Frames {.smaller}

Indexing lets you access specific parts of your data:

-   Use `[row, column]` notation to select specific cells
-   Use `$` to access columns by name
-   Leave row or column blank to select all

```{r, echo=TRUE, eval=FALSE}
print (YourFile[,2])
print (YourFile$temp)
print (YourFile[2,])
plot (YourFile$temp, YourFile$elevation)
```

## Indexing in data frames {.flexbox .vcenter}

-   You can perform operations on particular levels of a factor
-   Calculating the mean of the 'mixed' and 'gipps' levels of habitat.
-   Note that the first argument is the numerical column vector, and the second is the factor column vector.
-   The third is the operation. Reversing the first two does not work (the one below).

```{r, eval = FALSE, echo = TRUE}
tapply(YourFile$temp, YourFile$habitat, mean)
tapply(YourFile$temp, YourFile$habitat, var)
```

## Understanding Pipes {.smaller}

Pipes let you chain operations together, reading left-to-right instead of inside-out:

```{r, eval = FALSE, echo = TRUE}
# Without pipes (nested - hard to read)
round(mean(sqrt(c(1, 4, 9, 16))), 2)

# With base R pipe |> (R 4.1+)
c(1, 4, 9, 16) |> sqrt() |> mean() |> round(2)

# With magrittr pipe %>% (from tidyverse)
c(1, 4, 9, 16) %>% sqrt() %>% mean() %>% round(2)
```

::: callout-tip
Both `|>` and `%>%` do the same thing. You'll see both in online examples. We'll primarily use `|>` in this course.
:::

____________________________

## R INTERLUDE \| Some real transcriptomic data {.flexbox .vcenter}

-   Examine the data file
-   How many many rows and columns are there?
-   How many different variables are there?
-   What are the general types of variables?
-   Now let’s read the data file into R and analyze it
-   This exercise will help you get used to reading in and manipulating genomic data files
-   First off, remember to set your working directory to find your file correctly

## Some real transcriptomic data {.smaller}

```{r, eval = FALSE, echo = TRUE}
RNAseq_Data <- read.table('<name_of_file>', header=TRUE, sep=',')

print (RNAseq_Data)
head (RNAseq_Data)
tail (RNAseq_Data)

print (RNAseq_Data[,2])
print (RNAseq_Data[1,])
print (RNAseq_Data[1,2])
print (RNAseq_Data$ENSGACG00000000010)
print (RNAseq_Data$ENSGACG00000000010>45.0)
```

## Summary stats and figures {.smaller}

```{r, eval = FALSE, echo = TRUE}
summary1 <- summary(RNAseq_Data $ENSGACG00000000003)
print (summary1)

hist(RNAseq_Data $ENSGACG00000000003)
boxplot(RNAseq_Data$ENSGACG00000000003)
boxplot(RNAseq_Data$ENSGACG00000000003~RNAseq_Data$Population)
plot(RNAseq_Data $ENSGACG00000000003, RNAseq_Data$ENSGACG00000000003)

boxplot(RNAseq_Data $ENSGACG00000000003~RNAseq_Data$Treatment, 
        col = "red", ylab = "Expression Level", xlab = "Treatment level", 
        border ="orange", 
        main = "Boxplot of variation in gene expression across microbiota treatments")
```

# Reproducible workflows using Markdown

## `R scripts` and `Quarto Markdown files` {.flexbox .vcenter}

-   Often we want to write scripts that can just be run
-   We can also embed code in Markdown files that provide more annotations
-   https://quarto.org/docs/authoring/markdown-basics.html
-   You can insert `Rchunks` into `Quarto markdown` documents


## Markdown basics

-   a very simplified way for standard typesetting
-   simple markdown can be rendered in numerous different ways
-   Lists, codeblocks, images and more can all be inserted

## Inserting equations in markdown

``` latex
$$e=mc^2$$
```

$$e=mc^2$$

``` latex
$$\iint\limits_{a}^{b} f(x,y) \, dx \, dy$$
```

$$\iint\limits_{a}^{b} f(x,y) \, dx \, dy$$